---
title: "monocle3"
output: html_notebook
---
```{r}
library(Seurat)
library(scran)
library(data.table)
library(dplyr)
library(patchwork)
library(tidyverse)
library(MAST)
library(pals)

paths <- system("realpath ~/230113_seurat_intron/*/*/*non_tumour_removed.Rds", intern = T)
paths_split <- unlist(lapply(strsplit(paths, split = "/"), function(x) paste(x[6], x[7], sep = "_")))
paths2 <- system("realpath ~/240603_seurat_intron/CA0027/dura*/*non_tumour_removed_v4.Rds", intern = T)
paths_split2 <- unlist(lapply(strsplit(paths2, split = "/"), function(x) paste(x[6], x[7], sep = "_")))

paths <- c(paths, paths2)
paths_split <- c(paths_split, paths_split2)
paths <- setNames(paths,paths_split)

srt<- readRDS(paths[1])
DefaultAssay(srt) <- "RNA"
lib.median <- median(srt$nCount_RNA)
target_pseudocount <- 1
srt <- srt %>% 
  NormalizeData(scale.factor = lib.median / target_pseudocount) %>% 
  ScaleData(do.scale = F)

set.seed(100)
poisson_fit2 <- modelGeneVarByPoisson(srt@assays$RNA@counts, dispersion = 0.2, size.factors = target_pseudocount * srt$nCount_RNA / lib.median)
residuals2 <- poisson_fit2$total - poisson_fit2@metadata$trend(poisson_fit2$mean)
names(residuals2) <- rownames(srt)
top_genes2 <- rownames(srt)[order(residuals2, decreasing = TRUE)]

srt <- srt %>% 
  RunPCA(features = top_genes2[1:2000]) %>% 
  FindNeighbors(dims = 1:20) %>% 
  FindClusters(resolution = 1) %>% 
  RunUMAP(dims = 1:20)

clean_module <- readRDS("~/regulon/new_regulon/archetype/2024_06/clean_module.Rds")
srt <- AddModuleScore(srt, features  = clean_module)
names(clean_module)
colnames(srt@meta.data)[grep("Cluster", colnames(srt@meta.data))] <- paste(names(clean_module), "module")

DimPlot(srt, label = T)
```


```{r fig.height=10, fig.width=8}
FeaturePlot(srt, paste(names(clean_module), "module"))


```

```{r}
expression_matrix <- srt@assays$RNA@counts
cell_metadata <- srt@meta.data 
gene_annotaion <- as.data.frame(rownames(srt))
rownames(gene_annotaion) <- rownames(srt)
colnames(gene_annotaion) <- "gene_short_name"

cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotaion)
cds <- preprocess_cds(cds, num_dim = 20, use_genes = top_genes2[1:2000])
plot_pc_variance_explained(cds)
cds <- reduce_dimension(cds,max_components = 3,preprocess_method= "PCA")
```

\

```{r}
plot_cells(cds, color_cells_by = "seurat_clusters", label_leaves=FALSE,
           label_branch_points=FALSE, show_trajectory_graph = F,label_groups_by_cluster=FALSE)

```
```{r}
plot_cells(cds, genes = clean_module[[3]])
```


```{r}
cds <- cluster_cells(cds)
plot_cells(cds, label_leaves=FALSE,
           label_branch_points=FALSE, show_trajectory_graph = F)
plot_cells(cds, label_leaves=FALSE,color_cells_by="partition", group_cells_by="partition",
           label_branch_points=FALSE, show_trajectory_graph = F)
```


## Learn the trajectory graph

```{r}
cds <- learn_graph(cds)
plot_cells(cds,gene = "AR",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)
plot_cells(cds,
           label_groups_by_cluster=T,
           label_leaves=T,
           label_branch_points=T)
```

Each leaf, denoted by light gray circles, corresponds to a different outcome (i.e. cell fate) of the trajectory. Black circles indicate branch nodes, in which cells can travel to one of several outcomes.

we need to tell Monocle where the "beginning" of the biological process is. We do so by choosing regions of the graph that we mark as "roots" of the trajectory.

```{r}
cds <- as.cell_data_set(srt)
cds <- cluster_cells(cds)
cds <- learn_graph(cds, use_partition = F, close_loop = F)
plot_cells(cds,gene = "AR",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)
plot_cells(cds,
           label_groups_by_cluster=T,
           label_leaves=T,
           label_branch_points=T)
```

## pseudo-time calculation by manually picking root

```{r}
cds <- order_cells(cds)
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5)
```
```{r}
cds_pr_test_res <- graph_test(cds, neighbor_graph="principal_graph", cores=1)

```

Note that some of the cells are gray. This means they have infinite pseudotime, because they were not reachable from the root nodes that were picked.

## pseudo-time calculation by programatically choose root

but we don't have time bin info so not work for us!

```{r eval=FALSE, include=FALSE}
# a helper function to identify the root principal points:
get_earliest_principal_node <- function(cds, time_bin="130-170"){
  cell_ids <- which(colData(cds)[, "embryo.time.bin"] == time_bin)
  
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))
```

### Subset cells by branch

It is often useful to subset cells based on their branch in the trajectory. The function `choose_graph_segments` allows you to do so interactively.

```{r eval=FALSE, include=FALSE}
cds_sub <- choose_graph_segments(cds)
```

## 3D trajectories

```{r}
cds_3d <- reduce_dimension(cds, max_components = 3)
cds_3d <- cluster_cells(cds_3d)
cds_3d <- learn_graph(cds_3d)
cds_3d <- order_cells(cds_3d, root_cells = colnames(cds)[colData(cds)$seurat_clusters %in% c("7")] )

plot_cells_3d(cds_3d, color_cells_by="seurat_clusters")
plot_cells_3d(cds_3d, color_cells_by="cluster")
plot_cells_3d(cds_3d, color_cells_by="pseudotime")
```

## Differential analysis

There are two approaches for differential analysis in Monocle:

-   **Regression analysis**: using `fit_models()`, you can evaluate whether each gene depends on variables such as time, treatments, etc.

-   **Graph-autocorrelation analysis**: using `graph_test()`, you can find genes that vary over a trajectory or between clusters.

test with onco plot genelist first against cluster

```{r}
genelist_onco <- readRDS("~/genomic/oncoplot/genelist_onco.Rds")
cds_subset <- cds[rowData(cds)$gene_short_name %in% genelist_onco,]
gene_fits <- fit_models(cds_subset, model_formula_str = "~cluster")
fit_coefs <- coefficient_table(gene_fits)
fit_coefs
```

`coefficient_table()` tests whether each coefficient differs significantly from zero under the [Wald test](https://en.wikipedia.org/wiki/Wald_test). By default, `coefficient_table()` adjusts these p-values for multiple hypothesis testing using the method of [Benjamini and Hochberg](https://en.wikipedia.org/wiki/False_discovery_rate).

```{r}
cluster_terms <- fit_coefs %>% filter(term != "(Intercept)")
cluster_terms <- cluster_terms %>% filter (q_value < 0.05) %>% select(gene_short_name, term, q_value, estimate)
cluster_terms
```

365 out of 545 genelist

%\>% filter(gene_short_name %in% unique(cluster_terms\$gene_short_name)[1:30]

```{r}
#plot_genes_violin(cds_subset, group_cells_by="subclone", ncol=2) +
      #theme(axis.text.x=element_text(angle=45, hjust=1))
library(ggrepel)
library(ggplot2)
ggplot(data = cluster_terms %>% filter(abs(estimate)>1), aes(x = factor(term), y = estimate, colour = gene_short_name))+geom_line(aes(group = gene_short_name))+ theme(legend.position = "none")+
  geom_label_repel(aes(label = gene_short_name),
                  nudge_x = 1,
                  na.rm = TRUE)
```

## genes change as pseudotime function

If you'd like to rank the genes by effect size, sort this table by the `morans_I`column, which ranges from -1 to +1. A value of 0 indicates no effect, while +1 indicates perfect positive autocorrelation and suggests that nearby cells have very similar values of a gene's expression. Significant values much less than zero are generally rare.

Once again, we turn to `graph_test()`, this time passing it `neighbor_graph="principal_graph"`, which tells it to test whether cells at similar positions on the trajectory have correlated expression:

Moran's I is a measure of multi-directional and multi-dimensional spatial autocorrelation. The statistic tells you whether cells at nearby positions on a trajectory will have similar (or dissimilar) expression levels for the gene being tested.

```{r}
ciliated_cds_pr_test_res <- graph_test(cds, neighbor_graph="principal_graph", cores=4)
pr_deg_ids <- row.names(subset(ciliated_cds_pr_test_res[order(ciliated_cds_pr_test_res$morans_I, decreasing = T),], q_value < 0.05))

```

```{r}
plot_cells(cds, genes=c(head(pr_deg_ids, 9)),
           show_trajectory_graph=FALSE,
           label_cell_groups=FALSE,
           label_leaves=FALSE)
```

```{r}
plot_cells(cds, genes=c(head(pr_deg_ids, 9)),
           show_trajectory_graph=T,
           label_cell_groups=F,
           label_leaves=F)
```

```{r}
gene_module_df <- find_gene_modules(cds[pr_deg_ids,], resolution=c(10^seq(-6,-1)))

cluster_group_df <- tibble::tibble(cell=row.names(colData(cds)), 
                                cell_group=cds@clusters$UMAP$clusters)
agg_mat <- aggregate_gene_expression(cds, gene_module_df, cluster_group_df)
module_dendro <- hclust(dist(agg_mat))
gene_module_df$module <- factor(gene_module_df$module, 
                                levels = row.names(agg_mat)[module_dendro$order])
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
pheatmap::pheatmap(agg_mat,
                   scale="column", clustering_method="ward.D2")

```

```{r}
cluster_group_df <- tibble::tibble(cell=row.names(colData(cds)), 
                                cell_group=cds@colData$subclone)
agg_mat <- aggregate_gene_expression(cds, gene_module_df, cluster_group_df)
row.names(agg_mat) <- stringr::str_c("Module ", row.names(agg_mat))
pheatmap::pheatmap(agg_mat,
                   scale="column", clustering_method="ward.D2")
```

```{r}
plot_cells(cds,
           genes=gene_module_df,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)

plot_cells(cds,
           genes=gene_module_df %>% filter(module %in% c(1, 16, 10, 19)),
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)


```

cannot handle over 100 genes

```{r}
cds_subset <- cds[rowData(cds)$gene_short_name %in%head(pr_deg_ids, 9),  ]
plot_genes_in_pseudotime(cds_subset, color_cells_by = "subclone",
                         min_expr=0.5)

```

### Analyzing branches in single-cell trajectories

```{r eval=FALSE, include=FALSE}
cds_subset <- choose_cells(cds)
subset_pr_test_res <- graph_test(cds_subset, neighbor_graph="principal_graph", cores=4)
pr_deg_ids <- row.names(subset(subset_pr_test_res, q_value < 0.05))
gene_module_df <- find_gene_modules(cds_subset[pr_deg_ids,], resolution=0.001)
agg_mat <- aggregate_gene_expression(cds_subset, gene_module_df)
module_dendro <- hclust(dist(agg_mat))
gene_module_df$module <- factor(gene_module_df$module, 
                                levels = row.names(agg_mat)[module_dendro$order])

plot_cells(cds_subset,
           genes=gene_module_df,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)
```

```{r geting all the object}
library(monocle3)
library(dplyr)
tumor <- readRDS("~/seurat_objects_01_12_2022/CA0027/paraaortic_lymph_node_1/CA0027_paraaortic_lymph_node_1_srt_new_filtering_doublets_non_tumour_removed.Rds")
subcloneanno <- readRDS("~/subclone/new_filtering_subclone_anno.Rds")
subanno <- subcloneanno$CA0027_paraaortic_lymph_node_1
expression_matrix <- tumor@assays$RNA@data
cell_metadata <- tumor@meta.data %>% mutate(subclone = as.factor(subanno$subclone[match(rownames(tumor@meta.data), subanno$cell)]))
gene_annotaion <- as.data.frame(rownames(tumor))
rownames(gene_annotaion) <- rownames(tumor)
colnames(gene_annotaion) <- "gene_short_name"
cds <- new_cell_data_set(expression_matrix,
cell_metadata = cell_metadata,
gene_metadata = gene_annotaion)
cds <- preprocess_cds(cds, num_dim = 100)
plot_pc_variance_explained(cds)
cds <- reduce_dimension(cds)
cds <- cluster_cells(cds, resolution=1e-5)
cds <- learn_graph(cds)
cds <- order_cells(cds, root_cells = colnames(cds)[colData(cds)$seurat_clusters %in% c("7")])

cds_3d <- reduce_dimension(cds, max_components = 3)
cds_3d <- cluster_cells(cds_3d)
cds_3d <- learn_graph(cds_3d)
cds_3d <- order_cells(cds_3d, root_cells = colnames(cds)[colData(cds)$seurat_clusters %in% c("7")] )

genelist_onco <- readRDS("~/genomic/oncoplot/genelist_onco.Rds")
cds_subset <- cds[rowData(cds)$gene_short_name %in% genelist_onco,]
gene_fits <- fit_models(cds_subset, model_formula_str = "~cluster")
fit_coefs <- coefficient_table(gene_fits)
cluster_terms <- fit_coefs %>% filter(term != "(Intercept)")
cluster_terms <- cluster_terms %>% filter (q_value < 0.05) %>% select(gene_short_name, term, q_value, estimate)

ciliated_cds_pr_test_res <- graph_test(cds, neighbor_graph="principal_graph", cores=4)
pr_deg_ids <- row.names(subset(ciliated_cds_pr_test_res[order(ciliated_cds_pr_test_res$morans_I, decreasing = T),], q_value < 0.05))

gene_module_df <- find_gene_modules(cds[pr_deg_ids,], resolution=c(10^seq(-6,-1)))

cluster_group_df <- tibble::tibble(cell=row.names(colData(cds)), 
                                cell_group=cds@clusters$UMAP$clusters)
agg_mat <- aggregate_gene_expression(cds, gene_module_df, cluster_group_df)
module_dendro <- hclust(dist(agg_mat))
gene_module_df$module <- factor(gene_module_df$module, 
                                levels = row.names(agg_mat)[module_dendro$order])
```
