---
title: "external_data"
output: html_document
date: "2024-08-13"
---
```{r setup, include=FALSE}
setwd("~/CASCADEpaper/paper/External_data_objects")
library(Seurat)
library(DESeq2)
library(SummarizedExperiment)
library(limma) 
library(edgeR) 
library(readxl)
library(dplyr)
library(data.table)
library(ggrepel)
library(ggplot2)
library(ComplexHeatmap)
library(circlize)
library(pals)
library(RColorBrewer)
library(ggsci)
library(patchwork)
library(ggpubr)
clean_module <- readRDS("~/CASCADEpaper/paper/Fig5_archetype/clean_module.Rds")
names(clean_module) <- c("AR","Inflammation", "NE1","NE2", "Cycling","Glycolysis"  )
color_high <- setNames( c("dodgerblue3", brewer.dark2(6)[2:6]),names(clean_module))
```

```{r compare_exp}
compare_exp <- function(lcpm, gene, group){
  ind <- split(colnames(lcpm), group)
  df <- lapply(ind, function(x){
     if(class(gene) == "character"){
       exp <- colMeans(lcpm[gene [gene%in% rownames(lcpm)], x, drop = F])
       return(data.frame(exp = exp, sample = names(exp), gene = gene))
     }
     if(class(gene) == "list"){
       return(sapply(gene, function(y) return(colMeans(lcpm[y[y %in% rownames(lcpm)], x, drop = F]))) %>% as.data.frame()  %>% tibble::rownames_to_column("sample") %>%
                reshape2::melt(id.vars = "sample", variable.name = "gene", value.name = "exp"))
     }
  })
 df <- rbindlist(df, idcol = "group")
 df$group <- factor(df$group, levels = levels(group))
return(df)
}
source("~/CASCADEpaper/paper/cols.R") 
source("~/CASCADEpaper/paper/functions.R") 
```


# Molecular profiling stratifies diverse phenotypes of treatment-refractory metastatic castration-resistant prostate cancer
## GSE126078
```{r}
# setwd("~/CASCADEpaper/paper/External_data_objects/GSE126078")
counts <- read.table(gzfile("GSE126078/GSE126078_raw_counts_GRCh38.p13_NCBI.tsv.gz"), header = T)
# FPKM <- read.table(gzfile("GSE126078/GSE126078_norm_counts_FPKM_GRCh38.p13_NCBI.tsv.gz"), header = T)
geneanno <-readr::read_tsv(gzfile("GSE126078/Human.GRCh38.p13.annot.tsv.gz"))
mol_prof_specimen <- readRDS("~/CASCADEpaper/paper/External_data_objects/mol_prof_specimen.Rds")
counts[1:3, 1:3]
class(counts)
sampleid <- colnames(counts)[-1]
head(sampleid)
geneid <- geneanno$Symbol[match(counts$GeneID, geneanno$GeneID)]
head(geneid)

counts <- as.matrix(counts[, 2:ncol(counts)])
rownames(counts) <- geneid

sample_name <- list.files("GSE126078/raw")
ids <- sapply(strsplit(sample_name, split = "_"), "[", 1)
names <- sapply(strsplit(sample_name, split = "_"), function(x) paste(x[!grepl("GSM|processed|data.txt.gz",x)], collapse = "_"))
sample_name <- setNames(names, ids)
colnames(counts) <- sample_name[colnames(counts)]

counts <- counts[, rownames(mol_prof_specimen)]
x <- DGEList(counts = counts)
x$samples$group <- factor(mol_prof_specimen$Classification, levels = c("AR+NE-" , "ARlowNE-", "AR-NE-","AR+NE+" ,"AR-NE+"))
x$samples$site <-  sapply(strsplit(rownames(mol_prof_specimen), split = "_"), function(x) tail(x, 1))
unique(x$samples$group)

```

```{r}
keep.exprs <- filterByExpr(x, group= x$samples$group)
x <- x[keep.exprs, keep.lib.sizes=FALSE]
x <- calcNormFactors(x, method = "TMM")
lcpm <- cpm(x, log=TRUE)
saveRDS(x, "GSE126078/x.Rds")
saveRDS(lcpm, "GSE126078/lcpm.Rds")
```

```{r fig.height=4, fig.width=8.5}
df <- compare_exp(lcpm, group = x$samples$group, gene = clean_module)
df
g1 <- ggplot(df, aes(x = group, y = exp, fill = group))+ 
  geom_violin(aes( color = group))+ geom_boxplot(width=0.1, outlier.size = 2) + theme_classic(base_size = 18) + 
  facet_wrap(.~gene, nrow = 1)+ 
  scale_fill_jama()+scale_color_jama()+
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())+
  theme(legend.position = "bottom", text = element_text(size = 16), legend.title.position = "top",
        plot.margin = margin(t = 0, l = 0, r = 0, b = 0, "cm"), panel.spacing = unit(0.1, "pt"), 
        legend.margin = margin(-10, 0, 0, 0))+
  labs(x = "Group", y = "Expression", fill = "Group", color = "Group")
  # guides(fill=guide_legend(nrow=2,byrow=TRUE))
g2 <- ggplot(df, aes(x = gene, y = exp, fill = gene))+ 
  geom_violin(aes(color = gene))+ geom_boxplot(width=0.1, outlier.size = 2) + theme_classic(base_size = 18) +
   facet_wrap(.~group, nrow = 1)+ 
  scale_fill_manual(values = color_high, name = "module")+scale_color_manual(values = color_high, name = "module")+
   theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + labs(x = "module")+
  theme(legend.position = "bottom", text = element_text(size = 16), legend.title.position = "top",
        plot.margin = margin(t = 0, l = 0, r = 0, b = 0, "cm"), panel.spacing = unit(0.1, "pt"), 
        legend.margin = margin(-10, 0, 0, 0))+
  guides(fill = guide_legend(nrow = 1))
  # guides(fill=guide_legend(nrow=2,byrow=TRUE))
g2
g1
ggsave("~/CASCADEpaper/paper/External_data_objects/pdf/GSE126078.pdf", width = 8.5, height = 4)

```
```{r}
df %>% group_by(gene,group) %>% summarise(mean = mean(exp))
compare_means(exp~group, data = df, group.by = "gene", p.adjust.method = "BH")
```


```{r}
leg1 <- as_ggplot(get_legend(g1))
leg2 <- as_ggplot(get_legend(g2))

leg1 + leg2
ggsave("~/CASCADEpaper/paper/External_data_objects/pdf/GSE126078_leg.pdf", width = 20, height = 2, units = "cm")

```

```{r}
boxplotexp <- function(lcpm, group, gene){
  df <- compare_exp(lcpm, group = group, gene = gene)
  # g <- ggplot(df, aes(x = group, y = exp, fill = group))+ 
  #   geom_violin(aes( color = group))+ geom_boxplot(width=0.1, outlier.size = 2) + theme_bw() + 
  #   facet_wrap(.~gene)+ 
  #   scale_fill_jama()+scale_color_jama()+
  #   theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),strip.background=element_rect(fill="white"))+
  #   theme(legend.position = "bottom", text = element_text(size = 16), legend.title.position = "top",
  #         plot.margin = margin(t = 0, l = 0, r = 0, b = 0, "cm"), panel.spacing = unit(0.1, "pt"))+
  #   guides(fill=guide_legend(nrow=2,byrow=TRUE))
  # return(g)
  return(df)
}
genes <- c( "CACNA2D1","NFIB","HEPACAM2", "RAD51B","HDAC9", "FHIT", "PRIM2", "PLCL2")

f <- lapply(genes, function(gene) boxplotexp(lcpm, group = x$samples$group,gene))
f <- rbindlist(f)
f$gene <- factor(f$gene, levels = genes)
ggplot(f, aes(x = group, y = exp, fill = group))+ 
    geom_violin(aes( color = group))+ geom_boxplot(width=0.1, outlier.size = 2) + theme_bw() +
    facet_wrap(.~gene, ncol = 2)+
    scale_fill_jama()+scale_color_jama()+
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),strip.background=element_rect(fill="white"))+
    theme(text = element_text(size = 16), legend.title.position = "top",
          plot.margin = margin(t = 0, l = 0, r = 0, b = 0, "cm"), panel.spacing = unit(0.1, "pt"))
```

## GSE264598 ASCL1 KO mouse modle 
Neuroendocrine Differentiation in Prostate Cancer Requires ASCL1
https://aacrjournals.org/cancerres/article-abstract/84/21/3522/749267/Neuroendocrine-Differentiation-in-Prostate-Cancer?redirectedFrom=fulltext

Neuroendocrine Differentiation in Prostate Cancer Requires ASCL1
```{r}
counts <- fread("~/CASCADEpaper/paper/External_data_objects/GSE264598/GSE264598_sc_obj_sctran_combined_counts.tsv.gz")
meta <- fread("~/CASCADEpaper/paper/External_data_objects/GSE264598/GSE264598_sc_obj_sctran_combined_annot_df.tsv.gz")
counts <- as.matrix(counts)
counts[1:4, 1:4]
rownames(counts) <- counts[, 1]
counts <- counts[, 2:ncol(counts)]
storage.mode(counts) <- "numeric"
head(meta)
unique(meta$CellType)
rownames(meta) <- meta$CellID
```

```{r}
mssrt  <- CreateSeuratObject(counts = counts, meta.data = meta)
head(mssrt@meta.data)
mssrt[["percent.mt"]] <- PercentageFeatureSet(mssrt, pattern = "^mt-")

VlnPlot(mssrt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(mssrt, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(mssrt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```


```{r}
dim(mssrt)
mssrt <- subset(mssrt, subset = nFeature_RNA > 200 & nFeature_RNA < 9000 & percent.mt < 5)
#filter 2000 cells due to high mitochondria genes proportion
```

```{r}
mssrt <- NormalizeData(mssrt)
mssrt <- FindVariableFeatures(mssrt, selection.method = "vst", nfeatures = 2000)
mssrt <- ScaleData(mssrt, do.scale = F,do.center = T)
mssrt <- RunPCA(mssrt, features = VariableFeatures(object = mssrt))
ElbowPlot(mssrt)
```


```{r}
mssrt <- FindNeighbors(mssrt, dims = 1:20)
mssrt <- FindClusters(mssrt, resolution = 0.5)
mssrt <- RunUMAP(mssrt, dims = 1:20)
DimPlot(mssrt)
```
```{r}
saveRDS(mssrt, "GSE264598/mssrt.Rds")
```

```{r}
DimPlot(mssrt, group.by = "CellType")
```

```{r}
head(mssrt@meta.data)
DimPlot(mssrt, group.by = "SampleID")
f1 <- DimPlot(mssrt, group.by = "SampleID")
f1 <- coneraxes(f1)
ggsave("pdf/GSE264598.pdf", width = 5, height = 4)
```
```{r}
FeaturePlot(mssrt, c("Syp", "Chga", "Ascl1", "Insm1"))+plot_annotation(title = "NE")
FeaturePlot(mssrt, c("Ar", "Krt18", "Krt8", "Cd24a"))+plot_annotation(title = "Luminal")
FeaturePlot(mssrt, c("Krt5", "Trp63", "Krt14", "Krt17"))+plot_annotation(title = "Basal")
FeaturePlot(mssrt, c("Vim", "Cdn2", "Zab2"))+plot_annotation(title = "EMT")
FeaturePlot(mssrt, c("Sox2", "Sox1", "Foxa1"))+plot_annotation(title = "Stem")
FeaturePlot(mssrt, c("Mki67", "Top2a"))+plot_annotation(title = "proliferation")

```

```{r}
VlnPlot(mssrt, c("Ar", "Ascl1"), group.by = "SampleID")
p1 <- VlnPlot(mssrt,"Ar", group.by = "SampleID")
p2 <- VlnPlot(mssrt,"Ascl1", group.by = "SampleID")
p1/p2+plot_layout(axes = "collect", guides= "collect")
```


```{r}
df <- data.frame(Ar = mssrt@assays$RNA@data["Ar", ], Ascl1 = mssrt@assays$RNA@data["Ascl1", ])
df$SampleID <- mssrt$SampleID
df <- melt(df)
head(df)
df$class <- ifelse(grepl("RPMA", df$SampleID), "ASCL1 KO", "ASCL1 WT")

ggplot(df, aes(x = SampleID, y = value, fill = variable))+
  geom_violin(scale = "width", trim = TRUE, alpha = 0.7)+
  geom_jitter(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.9),
              size = 0.5, alpha = 0.6)+
  theme_classic(base_size = 18)+
  scale_fill_manual(values = c("steelblue1", "pink"), name = "gene")+
  theme(axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1),
        legend.margin = margin(l = -10), 
        )+
  labs(y = "Expression")+
  facet_grid(.~class, space = "free", scale = "free")
  

ggsave("pdf/GSE264598_ar_ascl.pdf", width = 5, height = 4)
```
```{r}
df %>% group_by(SampleID, variable) %>% summarise(avg = mean(value))
compare_means( value~ SampleID,  data = df[df$variable == "Ar", ], method = "wilcox")
compare_means( value~ SampleID,  data = df[df$variable == "Ascl1", ], method = "wilcox")
```


```{r include=FALSE}
mouse_human_genes = read.csv("http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt",sep="\t")
convert_human_to_mouse <- function(gene_list) {
  genes <- sapply(gene_list, function(gene){
      class_key = (mouse_human_genes %>% filter(Symbol == gene & Common.Organism.Name == "human"))[['DB.Class.Key']]
      if( !identical(class_key, integer(0)) ){
            return((mouse_human_genes %>% filter(DB.Class.Key == class_key & Common.Organism.Name=="mouse, laboratory"))[,"Symbol"])
      }else{
        print(paste("no match gene found for", gene))
        return(NULL)
        }
  })
  return(unlist(genes))
}


convert_mouse_to_human <- function(gene_list) {
  genes <- sapply(gene_list, function(gene){
      class_key = (mouse_human_genes %>% filter(Symbol == gene & Common.Organism.Name == "mouse, laboratory"))[['DB.Class.Key']]
      if( !identical(class_key, integer(0)) ){
            return((mouse_human_genes %>% filter(DB.Class.Key == class_key & Common.Organism.Name== "human"))[,"Symbol"])
      }else{print(paste("no match gene found for", gene))}
  })
  return(unlist(genes))
}

mu_module <- lapply(clean_module, convert_human_to_mouse)
```

```{r}
mssrt <- AddModuleScore(mssrt, mu_module)
f1 <- coneraxes(FeaturePlot(mssrt, "Cluster3")+ggtitle("NE1")+NoLegend()) 
f2 <- rmaxes(FeaturePlot(mssrt, "Cluster4")+ggtitle("NE2")+NoLegend()) 
f1+f2+plot_layout(ncol = 2)
ggsave("pdf/GSE264598.pdf", width = 9, height = 4, units = "cm")
```
```{r}
p1 <- FeatureScatter(mssrt, "Cluster3", "Cluster4", cells = WhichCells(mssrt, idents = levels(Idents(mssrt))[1]))

plist <- lapply(levels(Idents(mssrt)), function(i) FeatureScatter(mssrt, "Cluster3", "Cluster4", group.by = "SampleID", cells = WhichCells(mssrt, idents = i)))
  
wrap_plots(plist) + plot_layout(axes = "collect") & labs(x = "NE1", "NE2") & theme(legend.position = "bottom", legend.margin = margin(t = -10))

ggsave("pdf/GSE264598_featurescatter.pdf", width = 9, height = 7)
```


```{r}
# install.packages("ggridges")
library(ggridges)
VlnPlot(mssrt, c("Cluster3", "Cluster4"), group.by = "CellType")
VlnPlot(mssrt, "Cluster3", group.by = "CellType")

df <- rbind(data.frame(exp = mssrt$Cluster3, celltype = mssrt$CellType, module = "NE1"),
            data.frame(exp = mssrt$Cluster4, celltype = mssrt$CellType, module = "NE2"))
head(df)
g1 <- ggplot(df, aes( x = exp, fill= module, y = celltype))+ geom_density_ridges(alpha = 0.6)+theme_bw(base_size = 9)+
  theme(text = element_text(size = 9), 
        legend.position = "none")+
  scale_fill_manual(values = c("#7570B3", "#E7298A"))
# ggsave("pdf/GSE264598_vio_celltype.pdf", width= 11, height = 8, units = "cm")
g1
```
```{r fig.height=5, fig.width=9}
df <- rbind(data.frame(exp = mssrt$Cluster3, sample = mssrt$SampleID, module = "NE1"),
            data.frame(exp = mssrt$Cluster4, sample = mssrt$SampleID, module = "NE2"))
head(df)
df$class <- ifelse(grepl("RPMA", df$sample), "ASCL1 KO", "ASCL1 WT")
ggplot(df, aes( x = exp, fill= module, y = sample))+ geom_density_ridges(alpha = 0.6)+theme_minimal(base_size = 16)+
  theme(legend.position = "bottom", 
        axis.title.y = element_blank(), 
        legend.margin = margin(t = -15))+
  scale_fill_manual(values = c( "#7570B3", "#E7298A"))+
  facet_grid(class~., scales = "free", space = "free")+
  labs(x = "average expression")
  
ggsave("pdf/GSE264598_ridges.pdf", width= 4.5, height = 3)
```

```{r}
df %>% group_by(sample, module) %>% summarise(avg = mean(exp))
compare_means( exp~ sample,  data = df[df$module == "NE1", ], method = "wilcox")
compare_means( exp~ sample,  data = df[df$module == "NE2", ], method = "wilcox")
```

```{r fig.height=5, fig.width=9}
df <- data.frame(Ar = mssrt@assays$RNA@data["Ar", ], Ascl1 = mssrt@assays$RNA@data["Ascl1", ])
df$SampleID <- mssrt$SampleID
df <- melt(df)
head(df)
df$group <- "AR/ASCL1 exp"
df <- rbind(df, data.frame(SampleID = mssrt$SampleID, variable = "difference", 
                           value = mssrt$Cluster3 - mssrt$Cluster4,
                           # value = (mssrt$Cluster3 + abs(min(mssrt$Cluster4)))/abs(mssrt$Cluster4), 
                           group = "difference"))

df$class <- ifelse(grepl("RPMA", df$SampleID), "ASCL1 KO", "ASCL1 WT")

ggplot(df, aes(x = SampleID, y = value, fill = variable))+
  geom_violin(scale = "width", trim = TRUE, alpha = 0.7)+
  geom_jitter(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.9),
              size = 0.5, alpha = 0.6)+
  geom_boxplot(width = 0.3, position = position_dodge(width = 0.9), outliers = F)+
  theme_classic(base_size = 18)+
  scale_fill_manual(values = c("steelblue1", "pink", "grey"), name = "gene", limits = c("Ar", "Ascl1"))+
  theme(axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1),
        legend.margin = margin(l = -10), 
        )+
  labs(y = "Expression")+
  facet_grid(group~class,  scale = "free")

ggsave("pdf/GSE264598_vio.pdf", width= 4.5, height = 6)
```
```{r}
df[df$variable == "difference", ] %>% group_by(SampleID, variable) %>% summarise(avg = mean(value))
compare_means( value~ SampleID,  data = df[df$variable == "difference", ], method = "wilcox")

```
```{r}
df[df$variable == "ratio", ] %>% group_by(SampleID) %>% summarise(med = median(value))
```


```{r}
VlnPlot(mssrt, c("Ascl1", "Ar"), group.by = "SampleID")
VlnPlot(mssrt, c("Ascl1", "Ar"), group.by = "CellType")

```


change to scantransform method
```{r eval=FALSE, include=FALSE}
mssrt  <- CreateSeuratObject(counts = counts, meta.data = meta)
head(mssrt@meta.data)
mssrt[["percent.mt"]] <- PercentageFeatureSet(mssrt, pattern = "^mt-")
mssrt <- SCTransform(mssrt, vars.to.regress = "percent.mt", verbose = FALSE)#Error in .GetSeuratCompat() : could not find function ".GetSeuratCompat"
mssrt <- RunPCA(mssrt, verbose = FALSE)
ElbowPlot(mssrt)
```

```{r}
mssrt <- RunUMAP(mssrt, dims = 1:30, verbose = FALSE)

mssrt <- FindNeighbors(mssrt, dims = 1:30, verbose = FALSE)
mssrt <- FindClusters(mssrt, verbose = FALSE)
DimPlot(mssrt, label = TRUE)
```

## Prostate Cancer Genome Atlas
```{r}
# atlas <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Dynamic_prostate_cancer_transcriptome/pca_atlas_dataset/pcatlas_dataset_vst_normalized.txt")
# atlas <- readRDS("~/CASCADEpaper/paper/External_data_objects/pcatlas_dataset/atlas_vst_normalized.Rds")
# head(atlas_meta)
# unique(atlas_meta$Sample.Type)
# atlas[1:5, 1:5]
# ensembl.genes <- atlas[, 1]
# atlas <- atlas[, 2:ncol(atlas)]
# atlas <- as.matrix(atlas)
# library(EnsDb.Hsapiens.v86)
# 
# geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v86, keys= ensembl.genes, keytype = "GENEID", columns = c("SYMBOL","GENEID"))
# geneID <- match(ensembl.genes, geneIDs1$GENEID)
# geneID <- geneIDs1$SYMBOL[geneID]
# geneID[is.na(geneID)] <- ensembl.genes[which(is.na(geneID))]
# rownames(atlas) <- geneID
# saveRDS(atlas, "~/CASCADEpaper/paper/External_data_objects/pcatlas_dataset/atlas_vst_normalized.Rds")
atlas_meta <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Dynamic_prostate_cancer_transcriptome/pca_atlas_dataset/pcatlas_dataset_annotations.txt")
atlas <- readRDS("~/CASCADEpaper/paper/External_data_objects/pcatlas_dataset/atlas.Rds")

```

```{r mean expression}
names(clean_module)
df <- lapply(names(clean_module), function(ns){
  genes <- clean_module[[ns]]
  temp <- genes[genes %in% rownames(atlas)]
  df <- data.frame(entry = atlas_meta$Entry, class = atlas_meta$Sample.Type, exp  = colMeans(atlas[temp, ]), module = ns)
  df$class <- factor(df$class, levels = c("NORMAL", "PRIMARY", "CRPC", "NEPC"))
  print(ggplot(df, aes(x = class, y = exp, fill = class)) +
          geom_violin(alpha = 0.8)+
          geom_jitter(color="black", size=0.4, alpha=0.9)+
          geom_boxplot(width = 0.1, fill = "white")+theme_bw(base_size = 9)+ggtitle(ns))
  return(df)
})


```
```{r}
df <- rbindlist(df)
ggplot(df, aes(x = class, y = exp, fill = module)) + 
  geom_violin(alpha = 0.8,position = position_dodge(0.9))+
  geom_boxplot(position = position_dodge(0.9), width = 0.1)+
  theme_bw(base_size = 16)+
  scale_fill_manual(values = color_high)+
  ggtitle("prostate cancer atlas")+
  theme(legend.position = "bottom") +
  ylab("avearge expression")

```
```{r}
df <- rbind(df , data.frame(entry = atlas_meta$Entry,class = atlas_meta$Sample.Type, exp  = atlas["AR", ], module = "AR_gene"))
df <- rbind(df, data.frame(entry = atlas_meta$Entry,class = atlas_meta$Sample.Type, exp  = atlas["ASCL1", ], module = "ASCL1_gene"))
df$module <- factor(df$module, levels = c("AR", "AR_gene", names(clean_module)[c(2, 5,6 ,3, 4)], "ASCL1_gene"))
ggplot(df, aes(x = class, y = exp, fill = module)) + 
  geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module))+
  geom_boxplot(position = position_dodge(0.9), width = 0.1)+
  theme_bw(base_size = 16)+
  scale_fill_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  scale_color_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  ggtitle("prostate cancer atlas")+
  theme(legend.position = "bottom") +
  ylab("avearge expression")
```



```{r fig.height=5, fig.width=5}
ggplot(df %>% filter(module %in% c("AR_gene", "ASCL1_gene", "AR", "NE1", "NE2")), aes(x = class, y = exp, fill = module ,color = module)) + 
  geom_violin(alpha = 0.8,position = position_dodge(0.9))+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, color ="black", outlier.size = 0.3)+
  theme_bw(base_size = 18)+
  scale_fill_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  scale_color_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  ggtitle("prostate cancer atlas")+
  guides(fill = guide_legend(nrow = 2, ncol = 3))+
  theme(legend.title.position = "top", axis.title.x = element_blank(),
        legend.position = "bottom",
    legend.key.size = unit(0.3, "cm"),          # Make the keys larger
    legend.spacing.x = unit(0.1, "cm"),       # Increase horizontal spacing
    legend.spacing.y = unit(0.1, "cm"),
    legend.margin = margin(t = -10, r = 0, b = 0, l = 0))+        # Increase vertical spacing) +
  ylab("vst normalized expression")
ggsave("pdf/PCAtlas_NEmodule.pdf", width = 5, height = 6)
```

```{r eval=FALSE, include=FALSE}
ggpaired(df%>% filter(module %in% c( "NE1", "AR_gene") & class == "CRPC"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2)+xlab("CRPC") +ylab("expression")

ggpaired(df%>% filter(module %in% c( "NE2", "AR_gene") & class == "CRPC"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2)+xlab("CRPC") +ylab("expression")
ggpaired(df%>% filter(module %in% c( "NE1", "ASCL1_gene") & class == "CRPC"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2)+xlab("CRPC") +ylab("expression")
ggpaired(df%>% filter(module %in% c( "NE2", "ASCL1_gene") & class == "CRPC"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2)+xlab("CRPC") +ylab("expression")


ggpaired(df%>% filter(module %in% c( "NE1", "ASCL1_gene") & class == "NEPC"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2) +xlab("NEPC") +ylab("expression")
ggpaired(df%>% filter(module %in% c( "NE2", "ASCL1_gene") & class == "NEPC"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2) +xlab("NEPC") +ylab("expression")

ggpaired(df%>% filter(module %in% c( "NE1", "ASCL1_gene") & class == "PRIMARY"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2)+xlab("PRIMARY") +ylab("expression")
ggpaired(df%>% filter(module %in% c( "NE2", "ASCL1_gene") & class == "PRIMARY"), x = "module", y = "exp", color = "module", line.color = "gray", line.size = 0.2)+xlab("PRIMARY") +ylab("expression")

```

```{r}
head(df)
tb <- dcast(df, entry~module, value.var = "exp")
tb$class <- atlas_meta$Sample.Type[match(tb$entry, atlas_meta$Entry)]

tb
ggplot(tb, aes(x =NE1, y = ASCL1_gene, color = class))+geom_point() + theme_bw(base_size = 16)+ 
  facet_wrap(.~class)
ggplot(tb, aes(x =NE2, y = ASCL1_gene, color = class))+geom_point() + theme_bw(base_size = 16)+ 
  facet_wrap(.~class)
ggplot(tb, aes(x =NE1, y = NE2, color = class, alpha = ASCL1_gene))+geom_point() + theme_bw(base_size = 16) + 
  xlim(c(6.5,11.5)) + ylim(c(6.5, 11.5))+
  labs(alpha = "ASCL1 expression", title = "prostate cancer atlas")+
  scale_alpha_continuous()+
  facet_wrap(.~class)
tb$group <- "other"
tb$group[tb$ASCL1_gene > 10 & tb$AR_gene > 10] <- "DP"
tb$group[tb$ASCL1_gene < 7 & tb$AR_gene < 7] <-"DN" 
f <- ggplot(tb, aes(x =NE1, y = NE2, color = group, alpha = AR_gene))+geom_point() + theme_bw(base_size = 16) + 
  xlim(c(6.5,11.5)) + ylim(c(6.5, 11.5))+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  labs(alpha = "AR expression", title = "prostate cancer atlas")+
  scale_alpha_continuous()+
  scale_color_manual(values = c("steelblue", "tomato", "lightgray"))+
  facet_wrap(.~class, ncol = 2)
print(f)

ggplot(tb, aes(x =AR_gene, y = ASCL1_gene, color = group, alpha = NE1))+geom_point() + theme_bw(base_size = 16) + 
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  labs(alpha = "ASCL1 expression", title = "prostate cancer atlas")+
  scale_alpha_continuous()+
  facet_wrap(.~class)

temp <- rbind(data.frame(ASCL1_exp = tb$ASCL1_gene, AR_exp = tb$AR_gene, expression = tb$NE1, module = "NE1", class  = tb$class),
              data.frame(ASCL1_exp = tb$ASCL1_gene, AR_exp = tb$AR_gene, expression = tb$NE2, module = "NE2", class  = tb$class))
ggplot(temp, aes(x =ASCL1_exp, y = expression, color = module))+geom_point() + theme_bw(base_size = 16)+ 
  facet_wrap(.~class) + scale_color_manual(values = color_high)
ggplot(temp, aes(x =AR_exp, y = expression, color = module))+geom_point() + theme_bw(base_size = 16)+ 
  facet_wrap(.~class) + scale_color_manual(values = color_high)
```
```{r fig.height=6, fig.width=5}
ggplot(tb, aes(x =NE1, y = NE2, color = group, alpha = AR_gene))+geom_point() + theme_bw(base_size = 16) + 
  xlim(c(6.5,11.5)) + ylim(c(6.5, 11.5))+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  labs(alpha = "AR\nexpression")+
  scale_alpha_continuous()+
  scale_color_manual(values = c("steelblue", "tomato", "lightgray"))+
  facet_wrap(.~class, ncol = 1)+
  theme(legend.spacing.x = unit(0.1, "cm"),       # Increase horizontal spacing
    legend.spacing.y = unit(0.1, "cm"),
    legend.margin = margin(t = 0, r = 0, b = 0, l = -10),
    strip.background = element_rect(fill = "white", color = "black"))
ggsave("pdf/PCAtlas_NE.pdf", width = 5, height = 6)
```

```{r}
df$connect <-  df$entry %in% df$entry[df$module == "ASCL1_gene"& df$exp > 9]

ggplot(df %>% filter(module %in% c( "ASCL1_gene","NE1", "NE2")), aes(x = class, y = exp, fill = module)) + 
  geom_violin(alpha = 0.8,position = position_dodge(0.9),width = 1.1, color = "white")+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers.size = 0.1,outliers = T)+
  # geom_point(data = df %>% filter(module %in% c( "ASCL1_gene","NE1", "NE2") & connect == TRUE), 
  #            position = position_dodge(0.9), aes(color = module), size = 0.3)+
  theme_bw(base_size = 16)+
  scale_fill_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  # scale_color_manual(values = c("black", "black", "black"))+
  ggtitle("prostate cancer atlas")+
  theme(legend.position = "bottom") +
  ylab("avearge expression")
```
```{r}
df$connect <-  df$entry %in% df$entry[df$module == "AR_gene"& df$exp <10]

ggplot(df %>% filter(module %in% c( "AR_gene","NE1", "NE2")), aes(x = class, y = exp, fill = module)) + 
  geom_violin(alpha = 0.8,position = position_dodge(0.9),width = 1.1, color = "white")+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers.size = 0.2,outliers = F)+
  geom_point(data = df %>% filter(module %in% c( "AR_gene","NE1", "NE2") & connect == TRUE), 
             position = position_dodge(0.9), aes(color = module), size = 0.3)+
  theme_bw(base_size = 16)+
  scale_fill_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  scale_color_manual(values = c("black", "black", "black"))+
  ggtitle("prostate cancer atlas")+
  theme(legend.position = "bottom") +
  ylab("avearge expression")+ 
  facet_grid(.~connect, scales = "free", space = "free")
```


```{r}
ggplot(df %>% filter(module %in% c( "ASCL1_gene","NE1", "NE2")), aes(x = module, y = exp, fill = module)) + 
  geom_jitter(size = 0.5, aes(color = module),position = position_dodge(0.9), alpha = 0.5) +
  geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module), width = 1.3)+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers = F)+
  geom_line(aes(group = entry), color = "gray")+
  theme_bw(base_size = 16)+
  scale_fill_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  scale_color_manual(values = c(color_high, "AR_gene" = "lightblue", "ASCL1_gene" = "red"))+
  ggtitle("prostate cancer atlas")+
  theme(legend.position = "bottom") +
  ylab("avearge expression")
```
```{r}
head(tb)
nrow(tb)
library(ggridges)

temp <- data.frame(tb[, c("AR_gene", "NE1", "class")], module = "NE1")
colnames(temp) <- c("AR_exp", "exp", "class", "module")
# temp <- temp %>% arrange(desc(exp)) %>% slice(1:round(nrow(tb)/2))
          
temp2 <- data.frame(tb[, c("AR_gene", "NE2", "class")], module = "NE2")
colnames(temp2) <- c("AR_exp", "exp", "class", "module")
# temp2 <- temp2 %>% arrange(desc(exp)) %>% slice(1:round(nrow(tb)/2))

plotdf <- rbind(temp, temp2)


ggplot(plotdf, aes(x = AR_exp, y = exp, color = module))+geom_point()+theme_bw(base_size = 16) + scale_color_manual(values = c("#7570B3", "#E7298A"))
ggplot(plotdf, aes( x = exp, fill= module, y = module))+ geom_density_ridges(alpha = 0.6)+theme_bw(base_size = 16)+
  theme()+
  scale_fill_manual(values = c("#7570B3", "#E7298A"))

ggplot(tb, aes(x=AR_gene, y=NE1) ) +
  geom_hex() +
  theme_bw()
ggplot(tb, aes(x=AR_gene, y=NE2) ) +
  geom_hex() +
  theme_bw()

ggplot(plotdf, aes( x = AR_exp, y = exp, group = module, color = module))+ geom_line()+theme_bw(base_size = 16)+
theme()+
scale_color_manual(values = c("#7570B3", "#E7298A"))

ggplot(plotdf, aes( x = factor(AR_exp), y = exp, fill = module))+ geom_bar(stat = "identity")+theme_bw(base_size = 16)+
  theme()+
  scale_fill_manual(values = c("#7570B3", "#E7298A"))

```
theme(legend.spacing.x = unit(0.1, "cm"),       # Increase horizontal spacing
    legend.spacing.y = unit(0.1, "cm"),
    legend.margin = margin(t = 0, r = 0, b = 0, l = -10),
    strip.background = element_rect(fill = "white", color = "black"))+
```{r}
plotdf$ARlv <- ntile(plotdf$AR_exp, 10)

medians <- plotdf %>%filter(class == "CRPC") %>% 
  group_by(ARlv, module) %>%
  summarize(median_value = median(exp),.groups = "drop")
medians$group <- ifelse(medians$module == "NE1", 1, 2)
# Add a line connecting the medians
p <- ggplot(plotdf %>% filter(class == "CRPC"), aes(x = factor(ARlv), y = exp, fill = module)) + 
  geom_boxplot(position = position_dodge(width = 0.75)) + theme_bw(base_size = 16)+
  scale_fill_manual(values = c("#7570B3", "#E7298A"))
p + 
  geom_line(data = medians, aes(x = ARlv+ (group - 1.5) * 0.4, y = median_value, color = module)) +
  geom_point(data = medians, aes(x = ARlv+ (group - 1.5) * 0.4, y = median_value), color = "black", shape = 2)+
  scale_color_manual(values = c("#7570B3", "#E7298A"))+labs(x = "AR expression levels", y = "module expression") + 
  ggtitle("CRPC")
ggsave("pdf/PCAtlas_trend.pdf", width = 5, height = 6)
```
```{r change to NE1/NE2 ratio}
plotdf <- temp
head(plotdf)
plotdf$ratio <- temp$exp/temp2$exp
plotdf$ARlv <- ntile(plotdf$AR_exp, 10)
plotdf$class <- factor(plotdf$class, levels = c("NORMAL", "PRIMARY", "CRPC", "NEPC"))

ggplot(plotdf, aes( x = AR_exp, y = ratio)) +geom_point() +facet_grid(.~class, scales = "free")+ theme_bw(base_size = 16)+geom_smooth(method=lm)+
  labs(x = "AR expression", y = "NE1/NE2 ratio")+
  theme(strip.background = element_rect(fill = "white", color = "black"))

# p <- ggplot(plotdf %>% filter(class == "CRPC"), aes(x = factor(ARlv), y = ratio)) + 
#   geom_boxplot(position = position_dodge(width = 0.75), fill = "sandybrown") + theme_bw(base_size = 16)
# p + 
#   geom_line(data = medians, aes(x = ARlv+ (group - 1.5) * 0.4, y = median_value, color = module)) +
#   geom_point(data = medians, aes(x = ARlv+ (group - 1.5) * 0.4, y = median_value), color = "black", shape = 2)+
#   scale_color_manual(values = c("#7570B3", "#E7298A"))+labs(x = "AR expression levels", y = "module expression") + 
#   ggtitle("CRPC")
```
```{r}
temp <- data.frame(tb[, c("ASCL1_gene", "AR_gene", "NE1", "class")], module = "NE1")
colnames(temp) <- c("ASCL1_exp","AR_exp", "exp", "class", "module")
# temp <- temp %>% arrange(desc(exp)) %>% slice(1:round(nrow(tb)/2))
          
temp2 <- data.frame(tb[, c("ASCL1_gene", "AR_gene", "NE2", "class")], module = "NE2")
colnames(temp2) <- c("ASCL1_exp","AR_exp", "exp", "class", "module")
# temp2 <- temp2 %>% arrange(desc(exp)) %>% slice(1:round(nrow(tb)/2))
temp$ratio <- temp$exp/temp2$exp
temp2$ratio <- temp$exp/temp2$exp
plotdf <- rbind(temp, temp2)
plotdf <- plotdf[plotdf$class %in% c("CRPC", "NEPC"), ]

plotdf %>% group_by(class) %>% summarise(ARcor = cor(AR_exp, ratio, method = "spearman"), ASCL1cor = cor(ASCL1_exp, ratio, method = "spearman"),
                                         ARcorP = cor.test(AR_exp, ratio, method = "spearman")$p.value, ASCL1corP = cor.test(ASCL1_exp, ratio, method = "spearman")$p.value, xAR = max(AR_exp), xASCL = max(ASCL1_exp), y = max(ratio)) -> labdf 
labdf$y <- max(labdf$y)
```


```{r}
g1 <- ggplot(plotdf, aes( x = AR_exp, y = ratio)) +geom_point() +facet_grid(.~class, scales = "free")+ theme_bw(base_size = 18)+geom_smooth(method=lm)+
  labs(x = "AR expression", y = "NE1/NE2 ratio", title =  "Correlation with AR")+
  theme(strip.background = element_rect(fill = "white", color = "black"))+
  geom_text(data = labdf, aes( x = xAR, y = y, label = paste("rho =" , signif(ARcor, 2))), hjust = 1, size = 5) 
g1
```

```{r}
g2 <- ggplot(plotdf, aes( x = ASCL1_exp, y = ratio)) +geom_point() +facet_grid(.~class, scales = "free")+ theme_bw(base_size = 18)+geom_smooth(method=lm)+
  labs(x = "ASCL1 expression", y = "NE1/NE2 ratio",title = "Correlation with ASCL1")+
  theme(strip.background = element_rect(fill = "white", color = "black"))+
  geom_text(data = labdf, aes( x = xASCL, y = y, label = paste("rho =" , signif(ASCL1cor, 2))), hjust = 1, size = 5)
g2
```
```{r}
g1+g2+plot_layout(axes = "collect", nrow = 2)
ggsave("pdf/PCAtlas_trend_ratio.pdf", width = 6, height = 6)
```


```{r}
library(ggExtra)
library(patchwork)
head(tb)

g <- ggplot(tb %>% filter(class == "CRPC"), aes(x =NE1, y = NE2))+
  geom_point(aes(color = AR_gene, size = AR_gene)) + theme_bw(base_size = 16) + 
  xlim(c(6.5,11.5)) + ylim(c(6.5, 11.5))+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  labs(alpha = "AR\nexpression")+
  scale_color_gradient(low = "lightgrey", high = "mediumblue", name = "AR\ngene\nexpression")+
  scale_size_continuous(name = "AR\ngene\nexpression")
leg <- as_ggplot(get_legend(g))
g <- g + NoLegend()
g <- as_ggplot(ggMarginal(g, type="density"))
g + leg + plot_annotation(title = "CRPC")
```
```{r}
head(tb)
g <- ggplot(tb %>% filter(class == "PRIMARY"), aes(x =NE1, y = NE2))+
  geom_point(aes(color = AR_gene, size = AR_gene)) + theme_bw(base_size = 16) + 
  xlim(c(6.5,11.5)) + ylim(c(6.5, 11.5))+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  labs(alpha = "AR\nexpression")+
  scale_color_gradient(low = "lightgrey", high = "mediumblue", name = "AR\ngene\nexpression")+
  scale_size_continuous(name = "AR\ngene\nexpression")
leg <- as_ggplot(get_legend(g))
g <- g + NoLegend()
g <- as_ggplot(ggMarginal(g, type="density"))
g + leg + plot_annotation(title = "primary")
```
```{r}
g <- ggplot(tb %>% filter(class == "CRPC"), aes(x =NE1, y = NE2))+
  geom_point(aes(color = AR_gene, size = AR_gene)) + theme_bw(base_size = 16) + 
  xlim(c(6.5,11.5)) + ylim(c(6.5, 11.5))+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  labs(alpha = "AR\nexpression")+
  scale_color_gradient(low = "lightgrey", high = "mediumblue", name = "AR\ngene\nexpression")+
  scale_size_continuous(name = "AR\ngene\nexpression")
leg <- as_ggplot(get_legend(g))
g <- g + NoLegend()
g <- as_ggplot(ggMarginal(g, type="density"))
g + leg + plot_annotation(title = "CRPC") 
```

```{r}
head(tb)
g <- ggplot(tb %>% filter(class %in% c("CRPC", "NEPC")), aes(x =AR_gene, y = ASCL1_gene,color = NE1, size = NE1))+
  geom_point() + theme_bw(base_size = 16) + 
  facet_grid(.~class, space = "free", scales = "free")+
  theme( strip.background = element_rect(fill = "white", color = "black"))+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  scale_color_gradient(low = "snow2", high = "#7570B3")+
  scale_size_continuous(range = c(0.5, 4)) +
  labs(x = "AR gene expression", y = "ASCL1 gene expression")+ guides(size = "none")
g2 <- ggplot(tb%>% filter(class %in% c("CRPC", "NEPC")), aes(x =AR_gene, y = ASCL1_gene,color = NE2, size = NE2))+
  geom_point() + theme_bw(base_size = 16) + 
  facet_grid(.~class, space = "free", scales = "free")+
  # geom_smooth(method = "lm", color = "grey", se = FALSE)+
  scale_color_gradient(low = "snow2", high = "#E7298A")+
  scale_size_continuous(range = c(0.5, 4))+
  theme( strip.background = element_rect(fill = "white", color = "black"))+
  labs(x = "AR gene expression", y = "ASCL1 gene expression")+ guides(size = "none")

g/g2 + plot_layout(axes = "collect")
ggsave("pdf/PCAtlas_scatter.pdf", width = 5, height = 6)
```
```{r}

```


```{r}
temp <- data.frame(tb[, c("ASCL1_gene", "NE1", "class")], module = "NE1")
colnames(temp) <- c("ASCL1_exp", "exp", "class", "module")
          
temp2 <- data.frame(tb[, c("ASCL1_gene", "NE2", "class")], module = "NE2")
colnames(temp2) <- c("ASCL1_exp", "exp", "class", "module")

plotdf <- rbind(temp, temp2)
plotdf <- filter(plotdf, class %in% c("CRPC", "NEPC"))

ggplot(plotdf, aes(x = ASCL1_exp, y = exp, color = module))+geom_point()+theme_bw(base_size = 16) + scale_color_manual(values = c("#7570B3", "#E7298A"))
ggplot(plotdf, aes( x = ASCL1_exp, fill= module, y = module))+ geom_density_ridges(alpha = 0.4)+theme_bw(base_size = 16)+
  theme()+
  scale_fill_manual(values = c("#7570B3", "#E7298A"))
```


## Living Tumour Lab Dataset
```{r}
living_data <- readRDS("~/CASCADEpaper/paper/External_data_objects/living_data.Rds")
living_tumour_dataset <- readRDS("~/CASCADEpaper/paper/External_data_objects/living_tumour_dataset.Rds")
living_tumour_dataset <- living_tumour_dataset[, c(1, 4:8)] %>% distinct()
head(living_tumour_dataset)
head(living_data)

########------
# living_dataset <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Living_tumour_laboratory/Living_Tumour_Laboratory_samples.txt")
# further_info <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Living_tumour_laboratory/Living_Tumour_Laboratory_data_sheets.txt")
# living_GSE_samples <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Living_tumour_laboratory/Living_Tumour_Laboratory_patient_sample_key.txt")
# living_GSE_samples$File <- NA
# for(sample in living_GSE_samples$Sample_ID){
#   temp <- grep(sample, list.files("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Living_tumour_laboratory/"))
#   temp_name <- list.files("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Living_tumour_laboratory/")[temp]
#   living_GSE_samples$File[living_GSE_samples$Sample_ID == sample] <- paste("/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Living_tumour_laboratory/", temp_name, sep="")
# }

```
```{r}
df <- lapply(names(clean_module), function(ns){
  genes <- clean_module[[ns]]
  temp <- genes[genes %in% rownames(living_data)]
  df <- data.frame(living_tumour_dataset, 
                   exp  = colMeans(living_data[temp, ]), module = ns)
  
  print(ggplot(df, aes(x = Hormone_sensitivty, y = exp, fill = Hormone_sensitivty)) + 
          geom_boxplot()+
          theme_bw(base_size = 16)+ggtitle(ns)+
          theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top"))
  print(ggplot(df, aes(x = Pathology, y = exp, fill = Pathology)) + 
          geom_boxplot()+
          theme_bw(base_size = 16)+ggtitle(ns)+
          theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top"))+
          guides(fill = guide_legend(title.position = "top"))
  return(df)
})
```

```{r}
df <- rbindlist(df)
head(df)
unique(df$Hormone_sensitivty)
ggplot(df, aes(x = Hormone_sensitivty, y = exp, fill = module)) + 
  theme_bw(base_size = 16)+
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module))+
  geom_boxplot(position = position_dodge(0.9))+
  theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top")+
  scale_fill_manual(values = c(color_high))+
  scale_color_manual(values = c(color_high))+
  ggtitle("Living Tumour Lab Dataset")+
  xlim(c("Androgen-dependent", "Partially androgen-dependent", "Androgen-independent" ))+
  ylab("avearge expression")
```

```{r}
df$Hormone_sensitivty <- factor(df$Hormone_sensitivty, levels = c("Androgen-dependent", "Partially androgen-dependent", "Androgen-independent" ))

f1 <- ggplot(df %>% filter(module %in% c("NE1", "NE2") & !is.na(df$Hormone_sensitivty)), aes(x = module, y = exp, fill = Hormone_sensitivty)) + 
  theme_bw(base_size = 18)+
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module))+
  geom_boxplot(position = position_dodge(0.9))+
  theme(legend.title.position ="top")+
  scale_fill_manual(values = rev(c("#D51317FF","#EFD500FF","#95C11FFF" )))+
  ggtitle("Living Tumour Lab Dataset")+
  ylab("expression")+
  labs(fill = "Hormone sensitivity")

```
```{r}
subdf <- df %>% filter(module %in% c("NE1", "NE2") & !is.na(df$Hormone_sensitivty))
temp <- split(subdf, subdf$module)
subdf$ratio <- rep(temp[[1]]$exp/temp[[2]]$exp, 2)     
f2 <- ggplot(subdf, aes(x = Hormone_sensitivty, y = ratio, fill = Hormone_sensitivty)) + 
  theme_bw(base_size = 18)+
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module))+
  geom_boxplot(position = position_dodge(0.9))+
  theme(legend.title.position ="top", legend.position = "right",
        axis.text.x = element_blank())+
  scale_fill_manual(values = rev(c("#D51317FF","#EFD500FF","#95C11FFF" )))+
  ylab("NE1/NE2 ratio")+
  labs(fill = "Hormone sensitivity", subtitle = "JonckheereTerpstraTest\np value = 0.0174")
  
f1+f2 + plot_layout(guides = "collect") & theme(legend.position = "bottom")&guides(fill = guide_legend(nrow = 2))
ggsave("pdf/Living_hormone_JT_ratio.pdf", width = 12, height = 6)
```
```{r}
DescTools::JonckheereTerpstraTest(subdf$ratio, g = subdf$Hormone_sensitivty, alternative = "increasing", nperm = 5000)
```


```{r}
pathology_color <- setNames(ggsci::pal_simpsons()(6), unique(df$Pathology)[!is.na(unique(df$Pathology))])
pathology_color <- c(pathology_color, setNames(c("#7570B3", "#E7298A"), c("NE1", "NE2")))

subdf <- df %>% filter(module %in% c("NE1", "NE2") & !is.na(df$Pathology) & !is.na(df$Hormone_sensitivty))
temp <- split(subdf, subdf$module)
subdf$ratio <- rep(temp[[1]]$exp/temp[[2]]$exp, 2)   

p1 <- ggplot(subdf, aes(x = module, y = exp, fill = Pathology)) + 
  theme_classic(base_size = 18)+
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module))+
  geom_boxplot(position = position_dodge(0.9))+
  geom_jitter(position = position_dodge(0.9), aes(color = Pathology))+
  theme(legend.title.position ="top", legend.position = "bottom")+
  scale_fill_manual(values = pathology_color)+
  scale_color_manual(values = pathology_color)+
  ylab("expression")+
  facet_grid(.~Hormone_sensitivty, space = "free", scales = "free")
p1

p2 <- ggplot(subdf, aes(x = Pathology, y = ratio, fill = Pathology)) + 
  theme_classic(base_size = 18)+
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), aes(color = module))+
  geom_boxplot(position = position_dodge(0.9))+
  geom_jitter(position = position_dodge(0.9), aes(color = Pathology))+
  theme(legend.title.position ="top", legend.position = "bottom")+
  scale_fill_manual(values = pathology_color)+
  scale_color_manual(values = pathology_color)+
  ylab("NE1/NE2 ratio")+
  facet_grid(.~Hormone_sensitivty, space = "free", scales = "free")
p2
```


```{r}
p1/p2 + plot_layout(axes = "collect", guides = "collect")
ggsave("pdf/Living.pdf", width = 8, height = 8)
```

```{r}
head(temp)

# data <- fread("GSE41192_living_tumor/GSE41192_Xenograft_Expression_Normalized_Genespring-Export_Sept2012.txt.gz")

pathology_color <- setNames(ggsci::pal_simpsons()(6), unique(df$Pathology)[!is.na(unique(df$Pathology))])
pathology_color <- c(pathology_color, setNames(c("#7570B3", "#E7298A"), c("NE1", "NE2")))
samples <- c("LTL-310","LTL-310R","LTL-311","LTL-311-G8","LTL-313A","LTL-313B","LTL-313B-G5", "LTL-313B-G8","LTL-313BR","LTL-313C","LTL-313D","LTL-313H","LTL-313H-G7","LTL-313H-G16","LTL-313HR","LTL-331", "LTL-331R", "LTL-331R-G7", "LTL-352","LTL-352-G13",
             "LTL-370","LTL-412","LTL-412-G5", "LTL-418","LTL-418-G6")
  m = "NE1"
  temp <- df %>% filter(module == m)
  temp <- temp[!is.na(temp$Pathology), ]
  temp  <- temp %>% filter(temp$Patient %in% names(which(table(temp$Patient)>1 )))
  temp$Sample <- factor(temp$Sample, levels = samples)
  temp2 <- temp
  
    m = "NE2"
  temp <- df %>% filter(module == m)
  temp <- temp[!is.na(temp$Pathology), ]
  temp  <- temp %>% filter(temp$Patient %in% names(which(table(temp$Patient)>1 )))
  temp$Sample <- factor(temp$Sample, levels = samples)
  
  
  plotdf <- rbind(temp, temp2)
  plotdf$module <- factor(plotdf$module, levels = c(levels(plotdf$Pathology), "NE1", "NE2"))
  ggplot(plotdf, aes(x = Sample, y = exp)) + 
    geom_line(aes(group = module, color = module))+ 
    geom_point(aes(color = Pathology), size = 3)+
    theme_bw(base_size = 16)+
    facet_grid(.~Patient, scales = "free", space = "free")+NoLegend()+
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0), legend.position = "bottom", 
          strip.background = element_rect(fill = "white", color = "black"),
          legend.margin = margin(-30, 0, 0, 0),
          legend.spacing.y = unit(0.1, 'cm'), 
          legend.title.position = "top")+
    scale_color_manual(values = pathology_color)+
    guides(color = guide_legend(nrow = 2))+
    xlab("")+ylab("module expression")+
    labs(color = "             Pathology")+
    ylim(c(6, 9))
ggsave("pdf/living_trend.pdf", width = 10, height = 6)
```
```{r}
temp3 <- temp
temp3$exp <- temp2$exp/temp$exp
temp3$module <- "NE1/NE2 ratio"
plotdf <- rbind(plotdf, temp3)
plotdf$grid <- c(rep("expression", 36), rep("ratio", 18))
ggplot(plotdf, aes(x = Sample, y = exp)) + 
    geom_line(aes(group = module, color = module))+ 
    geom_point(aes(color = Pathology), size = 3)+
    theme_bw(base_size = 16)+
    facet_grid(grid~Patient, scales = "free")+
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "bottom", 
          strip.background = element_rect(fill = "white", color = "black"),
          legend.margin = margin(-30, 0, 0, 0),
          legend.spacing.y = unit(0.1, 'cm'), 
          legend.title.position = "top", 
          axis.title.y = element_blank())+
    scale_color_manual(values = pathology_color)+
    guides(color = guide_legend(nrow = 2))+
    xlab("")+
    labs(color = "             Pathology")
ggsave("pdf/living_trend_ratio_colorbypathogloy.pdf", width = 10, height = 6)
color2 <- setNames(c(c("#7570B3", "#E7298A"),rev(c("#D51317FF","#EFD500FF","#95C11FFF" ))), 
                   c("NE1", "NE2", "Androgen-dependent" ,"Partially androgen-dependent" ,"Androgen-independent"))
saveRDS(plotdf, "~/CASCADEpaper/paper/External_data_objects/GSE41192_living_tumor/plotdf.Rds")
unique(plotdf$Patient)
plotdf$Patient <- factor(plotdf$Patient, levels = c("LTL-310","LTL-311","LTL-412","LTL-418", "LTL-313B", "LTL-313H","LTL-331"))
ggplot(plotdf, aes(x = Sample, y = exp)) + 
    geom_line(aes(group = module, color = module))+ 
    geom_point(aes(color = Hormone_sensitivty), size = 3)+
    theme_bw(base_size = 18)+
    facet_grid(grid~Patient, scales = "free")+
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "bottom", 
          strip.background = element_rect(fill = "white", color = "black"),
          legend.margin = margin(-30, 0, 0, 0),
          legend.spacing.y = unit(0.1, 'cm'), 
          legend.title.position = "top", 
          axis.title.y = element_blank())+
    scale_color_manual(values =color2)+
    guides(color = guide_legend(nrow = 2))+
    xlab("")+
    labs(color = "             Hormone sensitivity")

ggsave("pdf/living_trend_ratio_colorbyhormone.pdf", width = 12, height = 6)
```

## GSE211856 xnograft drug resistance
```{r GSE211856 xnograft drug resistance}
data <- read.table(gzfile("~/CASCADEpaper/paper/External_data_objects/GSE211856/GSE211856_TPM_LNCAP.xenograft_ccnelson.txt.gz"), header = TRUE, sep = "\t")
head(data)
rownames(data) <- data[, 1]
data <- data[, 2:ncol(data)]
class <-  sapply(strsplit(colnames(data), ".", fixed = T), '[', 2)
class <- factor(class, levels = c("INTACT","POSTCX","CRPC" ,"ENZ"))
```
```{r}
my_comparisons <- list( c("INTACT", "POSTCX"), c("POSTCX", "CRPC"), c("CRPC", "ENZ") )
df <- lapply(names(clean_module), function(ns){
  genes <- clean_module[[ns]]
  temp <- genes[genes %in% rownames(data)]
  df <- data.frame(exp  = colMeans(data[temp, ]), module = ns, class = class, AR_exp = as.numeric(data["AR", ]), ASCL1_exp = as.numeric(data["ASCL1",]))
  
  print(ggboxplot(df, x = "class", y = "exp",
          color = "class", palette = "jco", title = ns)+ 
  stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = 0.2))
  return(df)
})

dfs <- rbindlist(df)
```
```{r}
temp1 <- df[[1]]
temp1$exp <- df[[3]]$exp/df[[4]]$exp
temp1$module <- "NE1/NE2 ratio"

ggplot(temp1, aes(x = class, y = exp, fill = class)) + 
          geom_violin(alpha = 0.8,position = position_dodge(0.9),width = 1.1, color = "white")+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers.size = 0.1,outliers = T)+
          # geom_boxplot()+ geom_jitter(width = 0.4, color = "grey")+
          theme_bw(base_size = 16)+ggtitle("NE1/NE2 ratio")+
          theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top")+
  labs(y = "NE1/NE2 ratio")
```
```{r}
my_comparisons <- list( c("INTACT", "POSTCX"), c("POSTCX", "CRPC"), c("CRPC", "ENZ") )
ggboxplot(temp1, x = "class", y = "exp",
          color = "class", palette = "jco")+ 
  stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = 0.2) 
```

```{r}




ggplot(df[[1]], aes(x = class, y = AR_exp, fill = class)) + 
          geom_violin(alpha = 0.8,position = position_dodge(0.9),width = 1.1, color = "white")+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers.size = 0.1,outliers = T)+
          # geom_boxplot()+ geom_jitter(width = 0.4, color = "grey")+
          theme_bw(base_size = 16)+ggtitle("AR gene expression")+
          theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top")
ggplot(df[[1]], aes(x = class, y = ASCL1_exp, fill = class)) + 
          geom_violin(alpha = 0.8,position = position_dodge(0.9),width = 1.1, color = "white")+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers.size = 0.1,outliers = T)+
          # geom_boxplot()+ geom_jitter(width = 0.4, color = "grey")+
          theme_bw(base_size = 16)+ggtitle("ASCL1 gene expression")+
          theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top")
```


## GSE137829
### patient 6
```{r}
##Patient 6
raw_counts_o<-read.delim(file="/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Luminal_neuro_transdifferentiation/GSE137829/GSM4711415_P6_gene_cell_exprs_table.txt")
duplicates <- c("ABCF2", "ATXN7", "COG8", "CYB561D2", "DIABLO", "EMG1", "HSPA14", "LINC01238", "MATR3", "POLR2J3", "RGS5", "SOD2", "TBCE", "TMSB15B")
raw_counts2 <- raw_counts_o[which(!raw_counts_o$Symbol %in% duplicates),]
rownames(raw_counts2) <- raw_counts2[,2]
raw_counts <- raw_counts2[, -c(1,2)]
patient6 <- CreateSeuratObject(counts = raw_counts, min.cells = 3, project = "GSE137829")
patient6[["percent.mt"]] <- PercentageFeatureSet(patient6, pattern = "^MT-")
patient6[["percent.rp"]] <- PercentageFeatureSet(patient6, pattern = "^RP")
VlnPlot(patient6, features = c("nFeature_RNA", "nCount_RNA", "percent.mt",
                               "percent.rp"), ncol = 4)
patient6_filtered <- subset(patient6, subset = nFeature_RNA > 200  & percent.mt < 10)

patient6_filtered <- NormalizeData(patient6_filtered)
patient6_filtered <- ScaleData(patient6_filtered)
patient6_filtered <- FindVariableFeatures(patient6_filtered)
patient6_filtered <- RunPCA(patient6_filtered)
patient6_filtered <- FindNeighbors(patient6_filtered)
patient6_filtered <- FindClusters(patient6_filtered)
patient6_filtered <- RunUMAP(patient6_filtered, dims=1:20)

DimPlot(patient6_filtered)

FeaturePlot(patient6_filtered, c("AR", "SYP", "CHGA", "ASCL1", "HDAC9"))
FeaturePlot(patient6_filtered, c("AR", "ASCL1", "CHGA"))
```

```{r}
patient6_filtered <- AddModuleScore(patient6_filtered, clean_module)
colnames(patient6_filtered@meta.data)[grep("Cluster", colnames(patient6_filtered@meta.data))] <- names(clean_module)

FeaturePlot(patient6_filtered, c("NE1", "NE2"))

```
```{r}
VlnPlot(patient6_filtered, "AR")
VlnPlot(patient6_filtered, "ASCL1")

plot <- FeaturePlot(patient6_filtered, "AR")
tumors <- CellSelector(plot)
patient6_tumor <- subset(patient6_tumor, cells = tumors)
```
```{r}
patient6_tumor <- NormalizeData(patient6_tumor) %>%  
  ScaleData() %>% 
  FindVariableFeatures()%>%
  RunPCA()%>% 
  FindNeighbors() %>% 
  FindClusters() %>% 
  RunUMAP(dims = 1:20)

DimPlot(patient6_tumor)
FeaturePlot(patient6_tumor, c("AR", "ASCL1"), blend = T, cols = c("lightgrey", "yellow", "blue"))
```
```{r}
p1 <- FeatureScatter(patient6_tumor, "AR", "ASCL1") + theme_minimal(base_size = 18) + labs(color = "Cluster")
p1
g1 <- coneraxes(FeaturePlot(patient6_tumor, "AR"))
g2 <- coneraxes(FeaturePlot(patient6_tumor, "ASCL1"))
g1+g2
```


```{r}
patient6_tumor <- AddModuleScore(patient6_tumor, clean_module)
f1 <- rmaxes(FeaturePlot(patient6_tumor, "Cluster3", cols = rev(brewer.rdylbu(10)))+ggtitle("NE1")+NoLegend()) 
f2 <- coneraxes(FeaturePlot(patient6_tumor, "Cluster4", cols = rev(brewer.rdylbu(10))) +ggtitle("NE2")+NoLegend())
p1+g1+g2+(f1/f2)+plot_layout(ncol = 4)
ggsave("pdf/GSE137829_p6.pdf", width = 18, height = 4)
```

### patient 4
```{r}
raw_counts_o<-read.delim(file="/trigos_team/CASCADE/ExternalDatasets/Public_prostate_single_cell/Luminal_neuro_transdifferentiation/GSE137829/GSM4089154_P4_gene_cell_exprs_table.txt")

duplicates <- c("ABCF2", "ATXN7", "COG8", "CYB561D2", "DIABLO", "EMG1", "HSPA14", "LINC01238", "MATR3", "POLR2J3", "RGS5", "SOD2", "TBCE", "TMSB15B")
raw_counts2 <- raw_counts_o[which(!duplicated(raw_counts_o$Symbol)),]
rownames(raw_counts2) <- raw_counts2[,2]
raw_counts <- raw_counts2[, -c(1,2)]
patient4 <- CreateSeuratObject(counts = raw_counts, min.cells = 3, project = "GSE137829")
patient4[["percent.mt"]] <- PercentageFeatureSet(patient4, pattern = "^MT-")
patient4[["percent.rp"]] <- PercentageFeatureSet(patient4, pattern = "^RP")
VlnPlot(patient4, features = c("nFeature_RNA", "nCount_RNA", "percent.mt",
                               "percent.rp"), ncol = 4)
patient4_filtered <- subset(patient4, subset = nFeature_RNA > 200  & percent.mt < 10)

patient4_filtered <- NormalizeData(patient4_filtered)
patient4_filtered <- ScaleData(patient4_filtered)
patient4_filtered <- FindVariableFeatures(patient4_filtered)
patient4_filtered <- RunPCA(patient4_filtered)
patient4_filtered <- FindNeighbors(patient4_filtered)
patient4_filtered <- FindClusters(patient4_filtered)
patient4_filtered <- RunUMAP(patient4_filtered, dims=1:20)

DimPlot(patient4_filtered)

FeaturePlot(patient4_filtered, c("AR", "SYP", "CHGA", "ASCL1", "HDAC9"))
FeaturePlot(patient4_filtered, c("AR", "ASCL1", "CHGA"))
```
```{r}
plot <- FeaturePlot(patient4_filtered, "AR")
tumors <- CellSelector(plot)
patient4_tumor <- subset(patient4_filtered, cells = tumors)
patient4_tumor <- NormalizeData(patient4_tumor) %>%  
  ScaleData() %>% 
  FindVariableFeatures()%>%
  RunPCA()%>% 
  FindNeighbors() %>% 
  FindClusters() %>% 
  RunUMAP(dims = 1:20)

DimPlot(patient4_tumor)
FeaturePlot(patient4_tumor, c("AR", "ASCL1"), blend = T, cols = c("lightgrey", "yellow", "blue"))
```


```{r}
p1 <- FeatureScatter(patient4_tumor, "AR", "ASCL1") + theme_minimal(base_size = 18) + labs(color = "Cluster")
p1
g1 <- coneraxes(FeaturePlot(patient4_tumor, "AR"))
g2 <- coneraxes(FeaturePlot(patient4_tumor, "ASCL1"))
g1+g2
patient4_tumor <- AddModuleScore(patient4_tumor, clean_module)
f1 <- rmaxes(FeaturePlot(patient4_tumor, "Cluster3", cols = rev(brewer.rdylbu(10)))+ggtitle("NE1")+NoLegend()) 
f2 <- coneraxes(FeaturePlot(patient4_tumor, "Cluster4", cols = rev(brewer.rdylbu(10)))+ggtitle("NE2")+NoLegend()) 
p1+g1+g2+(f1/f2)+plot_layout(ncol = 4)
ggsave("pdf/GSE137829_p4.pdf", width = 18, height = 4)
```


```{r}
saveRDS(patient4_filtered, "GSE137829/patient4_filtered.Rds")
saveRDS(patient4_tumor, "GSE137829/patient4_tumor.Rds")
```


## GSE92721
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92721
Roxane's dataset, as suggested by Luc

```{r}
exp <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Mouse_models/GSE92721_RNAseq_normalized.txt")
rownames(exp) <- exp$GeneName
exp$GeneName <- NULL
boxplot(exp)

sample_map <- read.delim("/trigos_team/CASCADE/ExternalDatasets/Mouse_models/Sample_map_GSE92721.txt", sep="\t")
sample_map$Mouse <- sapply(strsplit(sample_map$ID, "_"), function(x){ return(x[2]) })
sample_map$ID1 <- sapply(strsplit(sample_map$ID, "_"), function(x){ return(x[1]) })
sample_map$group <- paste(sample_map$Mouse, 
                          ifelse(grepl("castrated", sample_map$description), "CRPC", "INTACT"), 
                          ifelse(grepl("abiraterone", sample_map$description), "AB", "vehicle"))
sample_map$group[1:3] <- "NORMAL"
sample_map$group <- factor(sample_map$group, levels = unique(sample_map$group))

pca <- prcomp(exp)
pca <- as.data.frame(pca$rotation[,c(1:2)])
pca$Sample <- rownames(pca)
pca$Condition <- sample_map$description[match(pca$Sample, sample_map$ID1)]
pca$Mouse <- sample_map$Mouse[match(pca$Sample, sample_map$ID1)]

ggplot(pca, aes(x=PC1, y=PC2))+
  geom_point(aes(shape=Mouse, colour=Condition), size=3)
```
```{r}
df <- lapply(names(mu_module), function(ns){
  genes <- mu_module[[ns]]
  temp <- genes[genes %in% rownames(exp)]
  df <- data.frame(sample_map,exp  = colMeans(exp[temp, ]), module = ns, AR_exp = as.numeric(exp["Ar", ]), ASCL1_exp = as.numeric(exp["Ascl1",]))
  
  print(ggplot(df, aes(x = group, y = exp, fill = Mouse)) + 
          geom_violin(alpha = 0.8,position = position_dodge(0.9),width = 1.1, color = "white")+
  geom_boxplot(position = position_dodge(0.9), width = 0.1, outliers.size = 0.1,outliers = T)+
          # geom_boxplot()+ geom_jitter(width = 0.4, color = "grey")+
          theme_bw(base_size = 16)+ggtitle(ns)+
          theme(axis.text.x = element_text(angle = -45, hjust = 0),legend.title.position ="top"))
  return(df)
})
```
```{r}

dfs <- rbindlist(df)
dfs$group <- paste0(dfs$group, ifelse(grepl("outlier",dfs$description), "(outlier)", ""), 
                    ifelse(grepl("exp",dfs$description), "(exp)", ""))
dfs$group <-  factor(dfs$group, levels = unique(dfs$group))
ggplot(dfs, aes(x = module, y = exp, fill = group)) + 
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), color = "white")+
  geom_boxplot()+
  theme_bw(base_size = 16)+
  scale_fill_manual(values = ggsci::pal_cosmic()(9))+
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  ylab("avearge expression")+
  facet_wrap(.~Mouse)
```

```{r}
subdf <- dfs %>% filter(Mouse %in% c("NPp53")) %>% filter(module %in% c("NE1", "NE2")) %>% filter(!grepl("outlier", group))



subdf %>% group_by(module) %>% summarise(rho = cor(exp, ASCL1_exp, method = "spearman"), p.value = cor.test(exp, ASCL1_exp, method = "spearman")$p.value) %>% mutate(labs =  paste("rho:", round(rho, 2)), x = 5, y = 8.5) -> labdf

subdf$group <- gsub("NPp53 ", "", subdf$group)
subdf$group <- factor(subdf$group, levels = unique(subdf$group))
```

```{r}

g1 <- ggplot(subdf, aes(x = ASCL1_exp, y = exp)) + geom_point(aes(color = group)) + theme_classic(base_size = 16)+
  geom_smooth(method=lm)+
  facet_grid(Mouse~module)+
  geom_text( data = labdf, aes(label = labs, x = x, y = y), vjust = 1, hjust = 0, size = 6)+
  theme(legend.position = "top",legend.title = element_blank())+
  scale_color_manual(values = ggsci::pal_cosmic()(9)[2:5])+
  guides(color = guide_legend(nrow = 2))+
  labs(y = "mean expression")
g1
```
```{r}
temp <- split(subdf, subdf$module)
subdf$ratio <- rep(temp[[1]]$exp/temp[[2]]$exp, 2)

ggplot(subdf, aes(x = ASCL1_exp, y = ratio, color = group)) + geom_point() + theme_classic(base_size = 18)+
  facet_grid(Mouse~.)
ggplot(subdf, aes(x = group, y = AR_exp, fill = group)) + geom_boxplot() + theme_classic(base_size = 18)
ggplot(subdf, aes(x = group, y = ASCL1_exp, fill = group)) + geom_boxplot() + theme_classic(base_size = 18)

```

```{r}
g2 <- ggplot(subdf, aes(x = module, y = exp, fill = group)) + 
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), color = "white")+
  geom_boxplot()+
  theme_classic(base_size = 16)+
  scale_fill_manual(values = ggsci::pal_cosmic()(9)[2:5])+
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        legend.title = element_blank()) +
  ylab("mean expression")+
  facet_wrap(.~Mouse) + 
  guides(fill = guide_legend(nrow = 2))
g2
```
```{r}
g3 <- ggplot(subdf, aes(x = group, y = ratio, fill = group)) + 
  # geom_violin(alpha = 0.8,position = position_dodge(0.9), color = "white")+
  geom_boxplot()+
  theme_classic(base_size = 16)+
  scale_fill_manual(values = ggsci::pal_cosmic()(9)[2:5])+
  theme(legend.position = "none",
        axis.text.x = element_text(angle = -45, hjust = 0, vjust = 1), 
        axis.title.x = element_blank()) +
  ylab("NE1/NE2 ratio")
g3
```
```{r}
compare_means(ratio ~ group,  data = subdf,
               method = "wilcox") %>% View()
```

```{r fig.height=9, fig.width=4.5}

g1/g2/g3
ggsave("pdf/GSE92721.pdf", width = 4.5, height = 9)
```

