---
title: "barplots_correlation"
output: html_document
date: '2024-01-08'
---

```{r setup, include=FALSE}
library(tidyverse)
# library(ggplot2)
library(pals)
library(scales)
library(Seurat)
library(colorRamps)
library(ggpubr)
library(geomtextpath)
# library(ggpubr)
library(dplyr)
library(circlize)
library(patchwork)
library(data.table)

final_signature <- readRDS("~/Fig4-5_archetype/clean_module.Rds")
new_name <- c("AR","Inflammation", "NE1","NE2", "Cycling","Glycolysis"  )
names(final_signature) <- paste0("Module", 1:6)
meta <- readRDS("~/Fig4-5_archetype/5_integrated_final_module_meta.Rds")
cells <- apply(meta[, grep("Module", colnames(meta))], 2, function(x){
  bins <- cut(x, breaks = 10)
  group <- bins %in% levels(bins)[5:10]
  return(setNames(group, rownames(meta)))
})

colnames(cells) <- paste0(colnames(cells), "_group") 
meta <- cbind(meta, cells)
meta$group <- "low expression"
ind_high <- apply(meta[, paste0("Module", 1:6, "_group")], 1, function(x) if(sum(x) > 1) paste(names(final_signature)[x], collapse = "&"))

for(i in names(final_signature)){
  meta$group[meta[, paste0(i, "_group")]] <- paste(i, "high")
}
meta$group[!sapply(ind_high, is.null)] <- unlist(ind_high)

```



## barplots 
need to run the first chunk anyway
```{r including transition, fig.height=7, fig.width=18}

# med_cells <- apply(meta[, names(final_signature)], 2, function(x) {
#   bins<- cut(x, breaks = 10)
#   return( bins %in% levels(bins)[3:5])
#   })
# 
# ind_med <- apply(med_cells, 1, function(x) if(sum(x) > 1) paste(names(final_signature)[x], collapse = "&"))
# for(i in names(final_signature)){
#   meta$group[med_cells[, i]] <- paste(i, "medium")
# }
# meta$group[!sapply(ind_med, is.null)] <- unlist(ind_med)
# table(meta$group)


df <- meta %>% group_by(sample, group) %>% summarise(n = n()) %>% mutate(freq = n/sum(n))

df$patient <- gsub("00","",substr(df$sample, 1, 6))
df$group  <- gsub("_", "", df$group)
colnames(df) <- c("sample", "module", "n","freq", "patient")
df$pathology <- ifelse(df$patient %in% c("CA90", "CA46"), "NE", ifelse(df$patient %in% c("CA27", "CA58"), "Mixed", "AD"))

color_high <- setNames( c("dodgerblue3", brewer.dark2(6)[2:6]),names(final_signature))
# color_median <- sapply(color_high, function(x) colorRamp2(c(0, 1), c("white", x))(0.5))
show_col(c(color_high))

mixed <- sapply(unique(c(unlist(ind_high))), function(x){
  modules <- strsplit(x, split = "&")[[1]]
  if (length(modules) <5){
  colors <- c(color_high[modules], rep("white", 4-length(modules)))
  c1 <- colorRamp2(c(0, 1), colors[1:2])(0.5)
  c2 <- colorRamp2(c(0, 1), colors[3:4])(0.5)
  final_color <-  colorRamp2(c(0, 1), c(c1, c2))(0.5)
  return(final_color)} else{ return("black")}
})
names(color_high) <- paste(names(color_high),"high")
# names(color_median) <- paste(names(color_median), "medium")


df$module <- factor(df$module, levels = c(unique(sort(grep("high", df$module, value = T))), unique(grep("&", df$module, value = T)), "low expression"))



colors <- c(color_high, mixed, setNames("lightgrey", "low expression"))
names(colors) <- gsub("_", "", names(colors))
colors <- colors[levels(df$module)]

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
moduledf <- df

moduledf$labs <- paste0(signif(moduledf$freq*100, 2), "%")
moduledf$labs[moduledf$freq<0.03] <- NA
saveRDS(moduledf, "moduledf.Rds")
ggplot(moduledf, aes(x = sample, y = freq, fill =module)) + geom_bar(stat = "identity") + 
  geom_text(aes(label = labs),
            position = position_stack(vjust = 0.5),        # Position text in the middle of the bars
            color = "black",                               # Set text color
            size = 4)+
  ggh4x::facet_nested(.~patient, scales = "free", space = "free") +
  scale_fill_manual(values = as.character(colors)) + 
  ggh4x::facet_nested(.~pathology+patient, scales = "free", space = "free") + 
  theme_classic(base_size = 18)+
  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90), legend.position = "none", 
        axis.title.x = element_blank(), panel.spacing = unit(0, "cm"))+
  scale_x_discrete(breaks = df$sample, labels= paste(sites, sapply(strsplit(df$sample, split = "_"), function(x) tail(x, 1))))+
  labs(y = "Cell Proportion")

ggsave("pdf/all_expression_moudle.pdf", width = 18, height = 5)

```
###by sites 
```{r fig.height=5, fig.width=18}
moduledf$site <- sites
moduledf$patient <- gsub("CA", "", moduledf$patient)
p <- ggplot(moduledf, aes(x = sample, y = freq, fill =module)) + geom_bar(stat = "identity") + 
  geom_text(aes(label = labs),
            position = position_stack(vjust = 0.5),        # Position text in the middle of the bars
            color = "black",                               # Set text color
            size = 4)+
  scale_fill_manual(values = as.character(colors)) + 
  ggh4x::facet_nested(.~site+patient, scales = "free", space = "free") + 
  theme_classic(base_size = 18)+
  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90), legend.position = "none", 
        axis.title.x = element_blank(), panel.spacing = unit(0, "cm"))+
  scale_x_discrete(breaks = df$sample, labels= paste(sites, sapply(strsplit(df$sample, split = "_"), function(x) tail(x, 1))))+
  labs(y = "Cell Proportion")
p
ggsave("pdf/expression_moudle_bysite.pdf", width = 18, height = 5)
moduledf$patient <- paste0("CA", moduledf$patient)
```
compare cycling module proportion between pathology
```{r}

moduledf %>% filter(module == "Module5 high") %>% group_by(pathology) %>% summarise(mean = mean(freq))

my_comparisons <- list( c("AD", "NE"), c("AD", "Mixed"), c("Mixed", "NE") )
ggboxplot(moduledf, x = "pathology", y = "freq",
          color = "pathology", palette = "jco",add = "jitter")+ 
  stat_compare_means(comparisons = my_comparisons)
```



```{r}
moduledf %>% filter(module == "Module5 high") %>% pull(freq) %>% summary()
moduledf %>% filter(module == "Module4 high" & pathology!="NE") 
rankdf <- df %>% group_by(sample) %>% mutate(order = order(freq, decreasing = T)) %>% select(c("sample", "module", "order"))
table(rankdf$module, rankdf$order)
rankdf$test <- grepl("Module6", rankdf$module)|grepl("Module1", rankdf$module)
table(rankdf$test, rankdf$order)
```


```{r legends, fig.height=3, fig.width=18}
df <- moduledf
p1 <- ggplot(df[grepl("high|low expression", df$module), ], aes(x = sample, y = freq, fill =module)) + geom_bar(stat = "identity") + 
  facet_grid(.~patient, scales = "free", space = "free") +
  theme(legend.margin = margin(l = -20, r=-10), 
        #legend.spacing.x = unit(6,"inches"), 
        legend.key.size = unit(6, "inches"))+
  scale_fill_manual(values = colors, name = "Module", guide = guide_legend(nrow = 4), labels = c(paste(new_name, "high"),   "low expression")) + 
  facet_grid(.~pathology+patient, scales = "free", space = "free") + 
  theme_bw(base_size = 18)
lg1 <- as_ggplot(get_legend(p1))
lg1 <- lg1 + theme(plot.margin = margin(l = -20))

subdf <- df[grepl("&", df$module), ] %>% filter(freq > 0.01) #%>% group_by(sample) %>% arrange(desc(freq)) %>% slice(1:5)
labs <- names(colors)[grepl("&", names(colors))]
labs <- labs %>% gsub("Module1", "AR",.) %>% gsub("Module2", "inflammation",.)%>% gsub("Module3", "NE1",.)%>% gsub("Module4", "NE2",.)%>% gsub("Module5", "cycling",.)%>% gsub("Module6", "hypoxia",.)
p3 <- ggplot(subdf, aes(x = sample, y = freq, fill =module)) + geom_bar(stat = "identity") + 
  facet_grid(.~patient, scales = "free", space = "free") +
  scale_fill_manual(values = colors[grepl("&", names(colors))], guide = guide_legend(title = "Transitioning Module", ncol = 2), 
                    labels = labs) + 
  facet_grid(.~pathology+patient, scales = "free", space = "free") + 
  theme_bw(base_size = 18)+
  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90), 
        legend.margin = margin(r = 1))+
  guides(fill = guide_legend(nrow = 7))


lg3 <- as_ggplot(get_legend(p3))

lg1+lg3+plot_layout(widths = c(1,3))
ggsave("pdf/module_legend.pdf", width = 18, height = 3)
```


```{r fig.height=7, fig.width=18}
p/(lg1+lg3)+plot_layout(heights = c(5,2), widths = c(2, 3))
ggsave("pdf/expression_moudle_bysite.pdf", width = 18, height = 7)
```
### piechart for module proportion
```{r}
df <- meta %>% group_by(group) %>% summarise(n = n()) %>% mutate(freq = n/sum(n), labs = paste( signif(freq*100, 3), "%"))

df$group <- factor(df$group, levels = c(unique(sort(grep("high", df$group, value = T))), unique(grep("&", df$group, value = T)), "low expression"))

df <- df %>% 
  arrange(desc(group)) %>%
  mutate(ypos = cumsum(freq)- 0.5*freq )
df$labs <- paste(df$group, df$labs, sep =":")
df$labs[df$freq < 0.01] <- NA


pie(df$freq, labels = df$labs, col = colors[df$group], border = "white")

pdf("pdf/pie_chart_module.pdf", width = 10, height = 10)
pie(df$freq, labels = df$labs, col = colors[df$group], border = "white")
dev.off()
```


### high only
```{r, fig.height=6, fig.width=10}
df<- meta %>% filter(grepl("high", group))%>% group_by(sample, group) %>% summarise(n = n()) %>% mutate(freq = n/sum(n))

df$patient <- gsub("00","",substr(df$sample, 1, 6))
df$group  <- gsub("_", "", df$group)
colnames(df) <- c("sample", "module", "n","freq", "patient")
df$pathology <- ifelse(df$patient %in% c("CA90", "CA46"), "NE", ifelse(df$patient %in% c("CA27", "CA58"), "Mixed", "AD"))

df$module <- factor(df$module, levels = c(unique(sort(grep("high", df$module, value = T)))))
sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites
highdf <- df

ggplot(df, aes(x = sample, y = freq, fill =module)) + geom_bar(stat = "identity") + theme_classic(base_size = 18)+
  scale_fill_manual(values = color_high,name = "high expression", guide = guide_legend(nrow = 2), labels = c(paste(new_name, "high"),   "low expression")) + 
  ggh4x:: facet_nested(.~pathology+patient, scales = "free", space = "free") + 
  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90), legend.position = "bottom")+
  scale_x_discrete(breaks = df$sample, labels= df$site) +ggtitle("moderate to high expression of module")
ggsave("pdf/moderate_to_high_expression_moudle.pdf", width = 9, height = 6)
```
```{r}
ggplot(df, aes(x = sample, y = freq, fill =module)) + geom_bar(stat = "identity") + theme_classic(base_size = 18)+
  facet_wrap(.~site, scales = "free") +
  scale_fill_manual(values = color_high,name = "high expression", guide = guide_legend(nrow = 2), labels = c(paste(new_name, "high"),   "low expression")) + 
  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90),, legend.position = "bottom")+
  scale_x_discrete(breaks = df$sample, labels= df$patient) +ggtitle("moderate to high expression of module")
ggsave("pdf/by_tissue_moderate_to_high_expression_moudle.pdf", width = 9, height = 9)
```

## HeatMap
```{r fig.height=12, fig.width=8}
library(ComplexHeatmap)
head(meta)
```


```{r fig.height=12, fig.width=8}
expmat <- readRDS("~/integration/2024_06/tumor_only_logPAC.Rds")
expmat <- as.matrix(expmat[as.character(unlist(final_signature)), ])
cells <- rownames(meta)[grepl("high", meta$group)]
cells <- split(cells, grep("high", meta$group, value = T))
names(cells) <- gsub(" high", "",names(cells))



avgexp <- sapply(cells, function(x) rowMeans(expmat[, x]))
avgexp <- avgexp[!rownames(avgexp) %in% rownames(avgexp)[duplicated(rownames(avgexp))], ]
avgexp <- avgexp[unlist(final_signature), ]



markers <- c("KLK3", "KLK2", "PCAT1", "FOLH1", "PDE4D" , "PMEPA1", "ACACA","AR", 
             "ADAMTS3", "ALCAM", "HIF1A","STAT3", "ARHGAP", "ARHGEF", "RUNX1", 
             "SYT1","SLC18A1","ST18", "NFATC2",
             "KCNB2", "HDAC9", "HEPACAM2", "PLCL2", "CADM2", "CTNNA2", "FAT4",
             "TOP2A","CENPF","CENPE", "BRCA1","E2F3", "EZH2",   "ATAD","MELK",     
             "ENO1","P4HA1", "PGK1", "PLOD2")
# markers <-c()
# for( i in names(final_signature)){
#   markers <- c(markers, names(head(sort(avgexp[intersect(final_signature[[i]], rownames(avgexp)), i], decreasing = T), 5)))
# }
# markers <- unique(markers)
# 
# markers <- sapply(names(final_signature), function(i){
#   names(head(sort(avgexp[intersect(final_signature[[i]], rownames(avgexp)), i], decreasing = T), 10))
# }, simplify = F)
# saveRDS(markers, "heatmap_marker.Rds")

ind2 <- which(rownames(avgexp) %in% markers)
TF2 <- rownames(avgexp)[ind2]
ha2 = rowAnnotation(foo = anno_mark(at = ind2,
                                    labels = TF2, 
                                    labels_gp = gpar(fontsize = 16)))
colors <- colorRamp2(c(min(avgexp), (max(avgexp) - min(avgexp))*0.4, max(avgexp)), c("royalblue", "lightyellow", "orangered"))
row_split <- unlist(lapply(names(final_signature), function(x) setNames(rep(x, length(final_signature[[x]])), final_signature[[x]])))
row_split <- row_split[rownames(avgexp)]
avgexp[1:4, 1:4]

```


```{r fig.height=12, fig.width=8}
source("~/CASCADEpaper/paper/cols.R")
pdf("pdf/module_heatmap.pdf", width = 7, height =12)


ht <- Heatmap(avgexp, col = colors, 
        show_column_names = T, show_row_names = F, show_row_dend = F,  row_title = "module signature", row_title_side = "left",
        right_annotation = ha2, 
        left_annotation = rowAnnotation(module = row_split, col = list(module = module_cols),annotation_name_gp = gpar(fontsize = 14),
                                        show_legend = F),
        row_split = row_split,name = "RUVIII\nnormalized\nexpression", 
        cluster_row_slices = F, cluster_columns = F, column_order = order(colnames(avgexp)), 
        border = T, 
        row_title_gp = gpar(fontsize = 18),
        column_names_gp = gpar(fontsize = 18),  # Increase font size of column labels
  heatmap_legend_param = list(
    title_gp = gpar(fontsize = 18),    # Increase font size of legend title
    labels_gp = gpar(fontsize = 18)))   # Increase font size of legend labels)
draw(ht, annotation_legend_side = "bottom", merge_legend = T)

dev.off()
draw(ht, annotation_legend_side = "bottom", merge_legend = T)

```



## hexagon plot
```{r}
library(igraph)

adjustment <- function(x){
  100*x/sum(x)
}
df <- meta %>% group_by(group) %>% mutate(sumn = n())%>% group_by(sample, group) %>% mutate(n = n()) %>% mutate(freq = n/sumn)

df$patient <- gsub("00","",substr(df$sample, 1, 6))
df$group  <- gsub("_", "", df$group)
colnames(df) <- c("sample", "module", "n","freq", "patient")
df$pathology <- ifelse(df$patient %in% c("CA90", "CA46"), "NE", ifelse(df$patient %in% c("CA27", "CA58"), "Mixed", "AD"))


df <- moduledf
pdf("pdf/module_cell_composition_network.pdf", width =14, height = 10)
par(mfrow= c(7,5), mar = c(1, 1, 1, 1))
for (i in unique(df$sample)){
temp <- df %>% filter(sample == i)
nodes <- temp %>% filter(grepl("high", module))
nodes <- as.data.frame(nodes)
rownames(nodes) <- gsub(" high", "",nodes$module)
nodes$id <- rownames(nodes)
nodes <- nodes[, c("id","module", "sample","n", "freq", "patient", "pathology")]
#print(head(nodes))
if(any(!paste("Module", 1:6, sep = "_")%in% nodes$id)){
miss_module <- setdiff(paste0("Module", 1:6),nodes$id)
nodes <- rbind(nodes, data.frame(id  = miss_module, module = "miss", sample = i, n = 0, freq = 0.00001, patient = unique(nodes$patient), pathology = unique(nodes$pathology)))
}

links <- temp %>% filter(grepl("&", module))
links <- as.data.frame(links)
edges <- apply(links, 1, function(x) {
temp <- strsplit(as.character(x["module"]), split = " ")[[1]][1]
temp <- strsplit(temp, split= "&")[[1]]
temp  <- combn(temp, 2)
return(as.data.frame(t(temp)) %>% `colnames<-`(c("from", "to"))%>% mutate(weight = x["freq"], module = paste(from, to, sep = "&")))
})
edges <- rbindlist(edges)
edges <- edges %>% group_by(module) %>% mutate(weight_sum = sum(as.numeric(weight)))
edges <- edges[, c("from", "to", "weight_sum")] %>% distinct()
#print(head(edges))
net <- graph_from_data_frame(d = edges, vertices = nodes[order(nodes$id),], directed = F)
#net <- simplify(net, remove.multiple = F, remove.loops = T)
V(net)$size <- log(adjustment(V(net)$freq)+1)*10
V(net)$color <- colors[paste(names(V(net)),"high")]
E(net)$width <- log(adjustment(E(net)$weight_sum)+1)*5
E(net)$edge.color <- apply(edges, 1, function(x) colors[paste(x[1], x[2], sep = "&")])
l <- layout_in_circle(net)
plot(net, layout = l, edge.color = E(net)$edge.color, vertex.label = NA, main  = i)
print(i)

}
dev.off()

```



## correlation with key gene expression

```{r}

phenotype_meta <- readRDS("~/Fig6_PSMA/phenotype_meta.Rds")
phenotype_meta$pathology <- ifelse(phenotype_meta$patient %in% c("CA90", "CA46"), "NE", ifelse(phenotype_meta$patient %in% c("CA27", "CA58"), "Mixed", "AD"))

head(phenotype_meta)
head(meta)
labname <- setNames(new_name, paste0("Module", 1:6))
correlation_expression<-function(modulemeta, expressionmeta, marker, pathology, title){
  # cells <- modulemeta$pathology  == pathology
  # modulemeta <- modulemeta[cells, ]
  # expressionmeta <- expressionmeta[cells, ]
  coR <- sapply(paste0("Module", 1:6), function(x) cor.test(modulemeta[, x], expressionmeta[, marker], method = "spearman",exact=FALSE))
  print(coR["p.value", ])
  df <-lapply(paste0("Module", 1:6), function(x){
  data.frame(score = modulemeta[,x], module = x, expression = expressionmeta[, marker])
    })
  df <- rbindlist(df)
  coR <- setNames(round(as.numeric(coR[4, ]), 3), paste0("Module", 1:6))
  df$correlation <- coR[df$module]
  df$module <- labname[df$module]
  
  g <- ggplot(df, aes(x = expression, y = score, color = module))+
  geom_point(alpha = 0.3) + 
  geom_labelsmooth(aes(label = correlation), fill = "white",
                method = "lm", formula = y ~ x,
                size = 3, linewidth = 1, boxlinewidth = 0.4, color = "black") +
    facet_wrap(.~module)+
  scale_color_manual(values = module_cols2, name = "module\ncorrelation")+
  theme_classic(base_size = 18)  + theme(legend.position = "none")+
    labs(x = "Marker Expression", y= "Module\nExpression")
  print(g)
  return(g)
}

f1 <- correlation_expression(modulemeta = phenotype_meta, expressionmeta = phenotype_meta, "AR_exp")
# f2 <- correlation_expression(modulemeta = phenotype_meta, expressionmeta = phenotype_meta, "AR_exp", "Mixed", 
#                              title = "Module correlaiton with AR expression in Mixed samples")
f3 <- correlation_expression(modulemeta = phenotype_meta, expressionmeta = phenotype_meta, "ASCL1_exp")
f4 <- correlation_expression(modulemeta = phenotype_meta, expressionmeta = phenotype_meta, "FOLH1_exp")
# f5 <- correlation_expression(modulemeta = phenotype_meta, expressionmeta = phenotype_meta, "FOLH1_exp", "Mixed", 
#                              title = "Module correlaiton with FOLH1 expression in Mixed samples")
ggsave("pdf/key_gene_correlation_AR.pdf", plot= f1, width = 9, height = 4)
ggsave("pdf/key_gene_correlation_ASCL.pdf", plot= f3, width = 9, height = 4)
```
```{r}
coR <- sapply(paste0("Module", 1:6), function(x) cor.test(phenotype_meta[, x], phenotype_meta[, "AR_exp"], method = "spearman",exact=FALSE))
coR
coR <- sapply(paste0("Module", 1:6), function(x) cor.test(phenotype_meta[, x], phenotype_meta[, "ASCL1_exp"], method = "spearman",exact=FALSE))
coR
```

```{r}
signature_meta_tumor_only <- readRDS("~/Fig2_signature/signature_meta_tumor_only.Rds")
head(signature_meta_tumor_only)
head(phenotype_meta)
new_meta <- cbind(phenotype_meta, signature_meta_tumor_only[, 19:44])

corlist <- list()
for (m in paste0("Module", 1:6)){
  cor <- sapply(colnames(signature_meta_tumor_only)[19:44], function(s){
    cor.test(phenotype_meta[, m], signature_meta_tumor_only[, s], method = "spearman", exact = F)
  })
  corlist[[m]] <- cor
  write.table(cor, paste0("correlation_pubsig/", m, ".txt"), quote = F, sep = ",")
}
saveRDS(corlist, "corlist_Fig2Sig.Rds")
```




## expression of key genes 
```{r}
cells <- rownames(meta)[grepl("high", meta$group)]
cells <- split(cells, grep("high", meta$group, value = T))
names(cells) <- gsub(" high", "",names(cells))

cells <- apply(phenotype_meta[, grep("_group", colnames(phenotype_meta))],2, function(x){
    rownames(phenotype_meta)[x]
  })
names(cells) <- paste0("Module", 1:6)


marker_expression <- function(marker = "AR_exp", phenotype_meta){
df <- lapply(names(cells), function(x){
  phenotype_meta[cells[[x]], c(x, marker)] %>% `colnames<-`(c("expression", "marker")) %>% mutate(module = labname[x])
})
  df <- rbindlist(df, fill = T)
  df$module <- factor(df$module, levels = as.character(rev(df %>% group_by(module) %>% summarise(mean = mean(marker)) %>% arrange(mean) %>% pull(module))))
compare_means(marker~module, data = df)
  f <- ggplot(df, aes(x = module, y = marker, fill = module))  + 
    geom_jitter(width = 0.25, alpha = 0.6, color = "grey", size = 0.2)+
    geom_boxplot()+
    scale_fill_manual(values = module_cols2)+theme_minimal(base_size = 18) + 
    ylab("Expression") + ggtitle(marker) +
    theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1), 
          legend.position = "none", axis.title.x = element_blank())+
    ggtitle(gsub("_exp", "",marker))
  print(f)
  return(df)
}

g1 <- marker_expression(marker = "AR_exp", phenotype_meta)
ggsave("pdf/AR_module.pdf", width = 9, height = 4)
g2 <- marker_expression(marker = "ASCL1_exp", phenotype_meta)
ggsave("pdf/ASCL_module.pdf", width = 9, height = 4)
g3 <- marker_expression(marker = "FOLH1_exp", phenotype_meta)
```
```{r}
compare_means(expression~module, data = g1, method = "wilcox.test", p.adjust.method = "BH")
compare_means(expression~module, data = g2, method = "wilcox.test", p.adjust.method = "BH")
```



## correlation plots with normal cell composition


### cell proportion between samples with more/less cells expression 
```{r fig.height=3, fig.width=3}
normal3 <- readRDS("~/normal_cell_annotation/subtype/final_normalsrt.Rds")
normal3$cell_type <- ifelse(normal3$cell_anno%in% c("T cells","B cells","Plasma cells",  "Macrophages"), "immune cells", 
                            ifelse(normal3$cell_anno%in% c("Fibroblasts","Adipocytes","Pericytes"), "stromal cells", "others"))
normal3$pathology <- ifelse(normal3$patient %in% c("CA0090", "CA0046"), "NE", ifelse(normal3$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))

df <- normal3@meta.data %>% group_by(patient,site,sample, cell_anno) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))

df$pathology <- ifelse(df$patient %in% c("CA0090", "CA0046"), "NE", ifelse(df$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))
df$patient <- gsub("00", "", df$patient)
df$labs <- paste0(signif(df$freq*100, 2), "%")
df$labs[df$freq < 0.03] <- NA

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites


cells <- apply(meta[, grep("_group", colnames(meta))],2, function(x){
    rownames(meta)[x]
  })
names(cells) <- paste0("Module", 1:6)


cell_split <- apply(high_cells, 2, function(x) split(x, meta$sample))
cell_pop <- lapply(cell_split, lapply, function(x) sum(x)/length(x))
cell_pop <- lapply(cell_pop, function(x) unlist(x))
lapply(cell_pop, function(x) median(x))
cell_pop <- lapply(cell_pop, function(x) x >median(x))

df
  for(i in names(cell_pop)){
  df$module_group <-  cell_pop[[i]][df$sample]
 # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
 #  facet_grid(.~module_group, scales = "free", space = "free") + 
 #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
 #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
 # print(f)
 # 
 # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
 #  facet_grid(module_group~site, scales = "free", space = "free") + 
 #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
 #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
 # print(f)
 # 
 # comp <- compare_means(freq ~ module_group, data = df, group.by = "cell_id")
  for(m in unique(df$cell_anno)){
   p <- ggboxplot(df[df$cell_anno == m,], x = "module_group", y = "freq",
            color = "module_group", palette = c("blue", "red"),
            add = "jitter", 
            short.panel.labs = T) + stat_compare_means(label =  "p.signif", label.x = 1.5, label.y = 0.75)
  #p <- facet(p, ncol = 7,nrow = 2,  facet.by = "module_group", short.panel.labs = T)
  print(ggpar(p, title = paste(i, m), legend = "bottom", tickslab.x = F))
    
  }
}
```


```{r fig.height=6, fig.width=8}
pdf("normal_cell_prop.vs.moudle_group.pdf", width = 8, height = 6)
for(m in unique(df$cell_anno)){
  i = "Module1"
  df$module_group <-  cell_pop[[i]][df$sample]
     p <- ggboxplot(df[df$cell_anno == m,], x = "module_group", y = "freq",
              color = "module_group", palette = c( "blue","red" ),
              add = "jitter", title = i, 
              short.panel.labs = T) + stat_compare_means(label =  "p.signif", label.x = 1.5, label.y = 0.45)+
       NoLegend()
  for(i in names(cell_pop)[2:6]){
    df$module_group <- cell_pop[[i]][df$sample]
   # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
   #  facet_grid(.~module_group, scales = "free", space = "free") + 
   #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
   #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
   # print(f)
   # 
   # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
   #  facet_grid(module_group~site, scales = "free", space = "free") + 
   #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
   #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
   # print(f)
   # 
   # comp <- compare_means(freq ~ module_group, data = df, group.by = "cell_id")
     g <- ggboxplot(df[df$cell_anno == m,], x = "module_group", y = "freq",
              color = "module_group", palette = c( "blue","red"),
              add = "jitter", title = i, 
              short.panel.labs = T) + stat_compare_means(label =  "p.signif", label.x = 1.5, label.y = 0.45)+
       NoLegend()
    #p <- facet(p, ncol = 7,nrow = 2,  facet.by = "module_group", short.panel.labs = T)
    p<- p+g
    }
  print(p + plot_layout(guides = "collect", nrow = 2, ncol = 3, axis_titles = "collect")+plot_annotation(title = m))
  
}
dev.off()
```


```{r}
df <- normal3@meta.data %>% group_by(sample, cell_anno) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))
df$cell_type <- ifelse(df$cell_anno%in% c("T cell","B cell","Plasma cell",  "Naive cell", "Macrophage", "Erythroblast"), "immune cells", "stromal cells")
df$cell_anno <- factor(df$cell_anno, levels = c("T cell","B cell","Plasma cell",  "Macrophage", "Erythroblast", "Epithelial cell","Chondrocyte", "Fibroblast", "Myofibroblast","Lipofibroblast"  ,"Endothelial cell", "Hepatocyte", "Neuron"))
normaldf <- df

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites

high_cells  

high_cells <- tumor_only_meta_expression %>% mutate(cells = FOLH1 > 0.3*(max(FOLH1) - min(FOLH1))) %>% pull(cells)
cell_split <- split(high_cells, tumor_only_meta_expression$sample)
cell_pop <- lapply(cell_split, function(x) sum(x)/length(x))

cell_pop <-unlist(cell_pop)
#cell_pop <- cell_pop > median(cell_pop)

head(cell_pop)

  df$cell_prop <- cell_pop[df$sample]
  df$patient <- substr(df$sample, 1, 6)
  df <- df %>% filter(!patient %in% c("CA0046", "CA0090"))
  df$module_group <-  df$cell_prop > median(cell_pop[!grepl("CA0090|CA0046", names(cell_pop))])
 # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
 #  facet_grid(.~module_group, scales = "free", space = "free") + 
 #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
 #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
 # print(f)
 # 
 # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
 #  facet_grid(module_group~site, scales = "free", space = "free") + 
 #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
 #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
 # print(f)
 # 
 # comp <- compare_means(freq ~ module_group, data = df, group.by = "cell_id")
 p <- ggboxplot(df, x = "module_group", y = "freq",
          color = "module_group", palette = c("blue", "red"),
          add = "jitter", 
          short.panel.labs = T) + stat_compare_means(label =  "p.signif", label.x = 1.5, label.y = 0.6)
p <- facet(p, ncol = 7,nrow = 2,  facet.by = "cell_anno", short.panel.labs = T)
print(ggpar(p, title = "FOLH1", legend = "bottom", tickslab.x = F))







c("SLC8A1−AS1", "UHRF1", "TYMS", "IGLL1", "UBE2C", "TOP2A", "NUSAP1", "TUBA1B", "MKI67", "TUBB", "HIST1H4C", "H2AFZ", "STMN1", "SIPA1L1", "AC120193.1", "IGHD", "ADAM28", "IL4R", "COL19A1", "IGLC3", "IGLC2", "IGKC", "IGHG2", "IGHG3", "IGHG1", "IGHA1", "FKBP11", "SSR4", "JCHAIN", "LTB", "CD83", "HLA−DQB1", "HLA−DQA1", "BANK1", "MARCH1", "LINC00926", "COTL1", "ARHGAP24", "MS4A1", "NSMCE1", "NEIL1", "BCL7A", "RUBCNL", "HCK", "FAM129C", "TCL1A", "PCDH9", "ACSM3", "CCDC191", "RAG1", "CD9", "PSD3", "ERG", "VPREB1", "DNTT", "AKAP12", "LINC01013", "ARPP21")

```






```{r fig.height=8, fig.width=14}

design <- "
  1234
  1567
"

labs <- structure(paste(unique(submodule$moduleid), sep = ":", c("stree", "NE", "cycling", "inflammation", "hypoxia", "AR")), names = names(cell_pop))
for (n in unique(df$cell_anno)){
    p <-  wrap_elements(grid::textGrob(n))
for(i in names(cell_pop)){
  df$module_group <-  cell_pop[[i]][df$sample] 
  subdf <-df[df$cell_anno == n,]
 # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
 #  facet_grid(.~module_group, scales = "free", space = "free") + 
 #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
 #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
 # print(f)
 # 
 # f <- ggplot(df, aes(x = sample, y = freq, fill = cell_anno)) + geom_bar(stat = "identity") + 
 #  facet_grid(module_group~site, scales = "free", space = "free") + 
 #  theme(axis.text.x = element_text(hjust = 1, vjust = 0.5, angle = 90)) + 
 #  scale_fill_manual(values = brewer.accent(13)) + ggtitle(i)
 # print(f)
 # 
 # comp <- compare_means(freq ~ module_group, data = df, group.by = "cell_id")
  
  if( nrow(subdf) > 5){
 f <- ggboxplot(subdf, x = "module_group", y = "freq",
          color = "module_group", palette = c("blue", "red"),
          add = "jitter", 
          short.panel.labs = T, title = i) + stat_compare_means(label =  "p.signif", label.x = 1.5, label.y = 0.75, method = "t.test")+
   ggtitle(labs[i])
# p <- facet(p, ncol = 7,nrow = 2,  facet.by = "cell_anno", short.panel.labs = T)
# print(ggpar(p, title = paste(i, n), legend = "bottom", tickslab.x = F))}
 p <- p+f
  }
}
  print(p  +plot_layout(guides = "collect",design = design, axis_titles = "collect")&theme(legend.position='bottom'))
}
p
```

###immune diversity index
```{r fig.height=5, fig.width=9}
library(vegan)
df <- as.data.frame.matrix(table( normal3$sample,  normal3$cell_anno))
head(df)
df.immune <- df[,colnames(df) %in% c("T cells","B cells","Plasma cells","Macrophages")]
head(df.immune)
diversity_ind <- diversity(df.immune)

df <- as.data.frame(diversity_ind) %>% rownames_to_column("sample") %>% mutate(patient = substr(sample, 1, 6))
ggplot(df, aes(x = sample, y = diversity_ind, fill = patient)) + geom_bar(stat = "identity") + theme_minimal(base_size = 18)+
  scale_x_discrete(labels = xlab) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1))

df2 <- meta %>% group_by(sample) %>% summarise(avg1 = mean(Module1),
                                                 avg2 = mean(Module2),
                                                 avg3 = mean(Module3),
                                                 avg4 = mean(Module4),
                                                 avg5 = mean(Module5),
                                                 avg6 = mean(Module6)) %>% 
  mutate(tile1 = ordered(ntile(avg1, 5)),
         tile2 =  ordered(ntile(avg2, 5)),
         tile3 =  ordered(ntile(avg3, 5)),
         tile4 =  ordered(ntile(avg4, 5)),
         tile5 =  ordered(ntile(avg5, 5)),
         tile6 =  ordered(ntile(avg6, 5)))

temp1 <- reshape2::melt(df2[,1:7], id.vars = "sample", variable.name = "avg", value.name = "exp")
temp1 <- temp1 %>% group_by(avg) %>% mutate(tile = ordered(ntile(exp, 5)))
temp1$avg <- as.character(temp1$avg)
for (i in 1:6){
  temp1$avg[temp1$avg == paste0("avg", i)] <- new_name[i]
}
colnames(temp1) <- c("sample", "module", "exp", "tile")
temp  <- merge(df, temp1, by = "sample", all=T)

# submodule <- moduledf %>% filter(grepl("high", module))
# submodule$moduleid <- gsub(" high", "", submodule$module)
# submodule <- submodule %>% group_by(sample, moduleid) %>% mutate(sum = sum(freq))
# submodule$diversity <- diversity_ind[submodule$sample]
# submodule <- submodule[, c(1, 5:6, 8:10)] %>% distinct()
# submodule 
# submodule %>% group_by(moduleid) %>% 
coR <- sapply(new_name, function(x){
  test <- temp[temp$module == x, ] %>% distinct()
  cor.test(test$diversity_ind, test$exp, method = "spearman",exact=FALSE)
} )
print(coR["p.value", ])
pva <- setNames(signif(as.numeric(coR["p.value", ]), 2), new_name)
rho <- setNames(signif(as.numeric(coR[4, ]), 2), new_name)

temp$pvalue <- pva[temp$module]
temp$rho <- rho[temp$module]
temp$facet <- paste0(temp$module, "\n", temp$pvalue)
temp
 ggplot(temp, aes(x = exp, y = diversity_ind, color =  module))+
  geom_point(alpha = 0.3) + 
  geomtextpath::geom_labelsmooth(aes(label = rho), fill = "white",
                   method = "lm", formula = y ~ x,
                   size = 5, linewidth = 1, boxlinewidth = 0.4, color = "black") +
  scale_color_manual(values = module_cols2, name = "module", labels = new_name)+
  facet_wrap(.~facet, scales = "free")+
  theme_classic(base_size = 18) + labs(x = "Module Expression", y = "Shannon Diversity Index") 
 ggsave("pdf/immune_index_moduleexp.pdf", width = 9, height = 5)
```
```{r fig.height=5, fig.width=9}
tumour_cells <- meta %>% group_by(sample) %>% summarise(n= n()) %>% pull(n)
df3 <- meta %>% group_by(sample) %>% summarise(n1 = sum(Module1_group),
                                               n2 = sum(Module2_group),
                                               n3 = sum(Module3_group),
                                               n4 = sum(Module4_group),
                                               n5 = sum(Module5_group),
                                               n6 = sum(Module6_group))
df3 <- as.data.frame(df3)
rownames(df3) <- df3$sample
df3 <- df3[, 2:7]
df3 <- as.data.frame(apply(df3, 2, function(x) x/tumour_cells ))
df3 <- df3 %>% mutate(tile1 = ordered(ntile(n1, 5)),
                      tile2 =  ordered(ntile(n2, 5)),
                      tile3 =  ordered(ntile(n3, 5)),
                      tile4 =  ordered(ntile(n4, 5)),
                      tile5 =  ordered(ntile(n5, 5)),
                      tile6 =  ordered(ntile(n6, 5)))
df3$sample <- rownames(df3)
temp1 <- reshape2::melt(df3[,c(1:6, 13)], id.vars = "sample", variable.name = "n", value.name = "prop")
temp1 <- temp1 %>% group_by(n) %>% mutate(tile = ordered(ntile(prop, 5)))
temp1$n <- as.character(temp1$n)
for (i in 1:6){
  temp1$n[temp1$n == paste0("n", i)] <- new_name[i]
}
colnames(temp1) <- c("sample", "module", "prop", "tile")
temp  <- merge(df, temp1, by = "sample", all=T)

coR <- sapply(new_name, function(x){
  test <- temp[temp$module == x, ] %>% distinct()
  cor.test(test$diversity_ind, test$prop, method = "spearman",exact=FALSE)
} )
print(coR["p.value", ])
pva <- setNames(signif(as.numeric(coR["p.value", ]), 2), new_name)
rho <- setNames(signif(as.numeric(coR[4, ]), 2), new_name)

temp$pvalue <- pva[temp$module]
temp$rho <- rho[temp$module]
temp$facet <- paste0(temp$module, "\n", temp$pvalue)
temp
 ggplot(temp, aes(x = prop, y = diversity_ind, color =  module))+
  geom_point(alpha = 0.3) + 
  geomtextpath::geom_labelsmooth(aes(label = rho), fill = "white",
                   method = "lm", formula = y ~ x,
                   size = 5, linewidth = 1, boxlinewidth = 0.4, color = "black") +
  scale_color_manual(values = module_cols2, name = "module", labels = new_name)+
  facet_wrap(.~facet, scales = "free")+
  theme_classic(base_size = 18) + labs(x = "Module High Proportion", y = "Shannon Diversity Index") 
 ggsave("pdf/immune_index_moduleprop.pdf", width = 9, height = 5)
```




###stromal diversity index
```{r}
df <- as.data.frame.matrix(table( normal3$sample,  normal3$cell_anno))
head(df)
df.stroma <- df[,!colnames(df) %in% c("Epithelial cells","Endothelial cells","Lymphatic endothelial cells","EndoMT","Fibroblasts","Adipocytes","Pericytes")]
head(df.stroma)
diversity_ind <- diversity(df.stroma)
barplot(diversity_ind)
df <- as.data.frame(diversity_ind) %>% rownames_to_column("sample") %>% mutate(patient = substr(sample, 1, 6))
ggplot(df, aes(x = sample, y = diversity_ind, fill = patient)) + geom_bar(stat = "identity") + theme_minimal(base_size = 18)+
  scale_x_discrete(labels = xlab) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1))

df2 <- meta %>% group_by(sample) %>% summarise(avg1 = mean(Module1),
                                                 avg2 = mean(Module2),
                                                 avg3 = mean(Module3),
                                                 avg4 = mean(Module4),
                                                 avg5 = mean(Module5),
                                                 avg6 = mean(Module6)) %>% 
  mutate(tile1 = ordered(ntile(avg1, 5)),
         tile2 =  ordered(ntile(avg2, 5)),
         tile3 =  ordered(ntile(avg3, 5)),
         tile4 =  ordered(ntile(avg4, 5)),
         tile5 =  ordered(ntile(avg5, 5)),
         tile6 =  ordered(ntile(avg6, 5)))

temp1 <- reshape2::melt(df2[,1:7], id.vars = "sample", variable.name = "avg", value.name = "exp")
temp1 <- temp1 %>% group_by(avg) %>% mutate(tile = ordered(ntile(exp, 5)))
temp1$avg <- as.character(temp1$avg)
for (i in 1:6){
  temp1$avg[temp1$avg == paste0("avg", i)] <- new_name[i]
}
colnames(temp1) <- c("sample", "module", "exp", "tile")
temp  <- merge(df, temp1, by = "sample", all=T)

# submodule <- moduledf %>% filter(grepl("high", module))
# submodule$moduleid <- gsub(" high", "", submodule$module)
# submodule <- submodule %>% group_by(sample, moduleid) %>% mutate(sum = sum(freq))
# submodule$diversity <- diversity_ind[submodule$sample]
# submodule <- submodule[, c(1, 5:6, 8:10)] %>% distinct()
# submodule 
# submodule %>% group_by(moduleid) %>% 
coR <- sapply(new_name, function(x){
  test <- temp[temp$module == x, ] %>% distinct()
  cor.test(test$diversity_ind, test$exp, method = "spearman",exact=FALSE)
} )
print(coR["p.value", ])
pva <- setNames(signif(as.numeric(coR["p.value", ]), 2), new_name)
rho <- setNames(signif(as.numeric(coR[4, ]), 2), new_name)

temp$pvalue <- pva[temp$module]
temp$rho <- rho[temp$module]
temp$facet <- paste0(temp$module, "\n", temp$pvalue)
temp
 ggplot(temp, aes(x = exp, y = diversity_ind, color =  module))+
  geom_point(alpha = 0.3) + 
  geomtextpath::geom_labelsmooth(aes(label = rho), fill = "white",
                   method = "lm", formula = y ~ x,
                   size = 5, linewidth = 1, boxlinewidth = 0.4, color = "black") +
  scale_color_manual(values = module_cols2, name = "module", labels = new_name)+
  facet_wrap(.~facet, scales = "free")+
  theme_classic(base_size = 18) + labs(x = "Module Expression", y = "Shannon Diversity Index") 
 ggsave("pdf/stroma_index_moduleexp.pdf", width = 9, height = 5)
 
 
tumour_cells <- meta %>% group_by(sample) %>% summarise(n= n()) %>% pull(n)
df3 <- meta %>% group_by(sample) %>% summarise(n1 = sum(Module1_group),
                                               n2 = sum(Module2_group),
                                               n3 = sum(Module3_group),
                                               n4 = sum(Module4_group),
                                               n5 = sum(Module5_group),
                                               n6 = sum(Module6_group))
df3 <- as.data.frame(df3)
rownames(df3) <- df3$sample
df3 <- df3[, 2:7]
df3 <- as.data.frame(apply(df3, 2, function(x) x/tumour_cells ))
df3 <- df3 %>% mutate(tile1 = ordered(ntile(n1, 5)),
                      tile2 =  ordered(ntile(n2, 5)),
                      tile3 =  ordered(ntile(n3, 5)),
                      tile4 =  ordered(ntile(n4, 5)),
                      tile5 =  ordered(ntile(n5, 5)),
                      tile6 =  ordered(ntile(n6, 5)))
df3$sample <- rownames(df3)
temp1 <- reshape2::melt(df3[,c(1:6, 13)], id.vars = "sample", variable.name = "n", value.name = "prop")
temp1 <- temp1 %>% group_by(n) %>% mutate(tile = ordered(ntile(prop, 5)))
temp1$n <- as.character(temp1$n)
for (i in 1:6){
  temp1$n[temp1$n == paste0("n", i)] <- new_name[i]
}
colnames(temp1) <- c("sample", "module", "prop", "tile")
temp  <- merge(df, temp1, by = "sample", all=T)

coR <- sapply(new_name, function(x){
  test <- temp[temp$module == x, ] %>% distinct()
  cor.test(test$diversity_ind, test$prop, method = "spearman",exact=FALSE)
} )
print(coR["p.value", ])
pva <- setNames(signif(as.numeric(coR["p.value", ]), 2), new_name)
rho <- setNames(signif(as.numeric(coR[4, ]), 2), new_name)

temp$pvalue <- pva[temp$module]
temp$rho <- rho[temp$module]
temp$facet <- paste0(temp$module, "\n", temp$pvalue)
temp
 ggplot(temp, aes(x = prop, y = diversity_ind, color =  module))+
  geom_point(alpha = 0.3) + 
  geomtextpath::geom_labelsmooth(aes(label = rho), fill = "white",
                   method = "lm", formula = y ~ x,
                   size = 5, linewidth = 1, boxlinewidth = 0.4, color = "black") +
  scale_color_manual(values = module_cols2, name = "module", labels = new_name)+
  facet_wrap(.~facet, scales = "free")+
  theme_classic(base_size = 18) + labs(x = "Module High Proportion", y = "Shannon Diversity Index") 
 ggsave("pdf/stroma_index_moduleprop.pdf", width = 9, height = 5)
```

### correlate module high proportion and normal cell proportion
```{r}
normal3 <- readRDS("~/normal_cell_annotation/subtype/final_normalsrt.Rds")
normal3$cell_type <- ifelse(normal3$cell_anno%in% c("T cells","B cells","Plasma cells",  "Macrophages"), "immune cells", 
                            ifelse(normal3$cell_anno%in% c("Fibroblasts","Adipocytes","Pericytes"), "stromal cells", "others"))
normal3$pathology <- ifelse(normal3$patient %in% c("CA0090", "CA0046"), "NE", ifelse(normal3$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))

df <- normal3@meta.data %>% group_by(patient,site,sample, cell_anno) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))

df$pathology <- ifelse(df$patient %in% c("CA0090", "CA0046"), "NE", ifelse(df$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))
df$patient <- gsub("00", "", df$patient)
df$labs <- paste0(signif(df$freq*100, 2), "%")
df$labs[df$freq < 0.03] <- NA

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites
df <- df %>% filter(n >10)

head(moduledf)
tumour_cells <- meta %>% group_by(sample) %>% summarise(n= n()) %>% pull(n)
df2 <- meta %>% group_by(sample) %>% summarise(n1 = sum(Module1_group),
                                               n2 = sum(Module2_group),
                                               n3 = sum(Module3_group),
                                               n4 = sum(Module4_group),
                                               n5 = sum(Module5_group),
                                               n6 = sum(Module6_group))
df2 <- as.data.frame(df2)
rownames(df2) <- df2$sample
df2 <- df2[, 2:7]
df2 <- as.data.frame(apply(df2, 2, function(x) x/tumour_cells ))
df2$sample <- rownames(df2)
temp1 <- sapply(df2[, c(paste0("n", 1:6))], function(x) ordered(ntile(x, 5)))
temp1 <- as.data.frame(temp1) %>% `colnames<-`(paste0("tile", 1:6))
df2 <- cbind(df2, temp1)
temp  <- merge(df, df2, by = "sample", all.x = T, all.y = F)


plotdf <- data.frame()
pdf("pdf/normalcell_propVSbinedmoduleprop_n=5.pdf", width = 10, height = 5)
for (n in levels(normal3$cell_anno)){
  subdf <-temp[temp$cell_anno == n,]
  for (m in 1:6){
    if (n_distinct(subdf[, paste0("tile", m )])==5 & all(table(subdf[, paste0("tile", m )])>=3)){
    test <- DescTools::JonckheereTerpstraTest(subdf$freq, g = subdf[, paste0("tile", m )], alternative = "increasing", nperm = 5000)
    test2 <- DescTools::JonckheereTerpstraTest(subdf$freq, g = subdf[, paste0("tile", m )], alternative = "decreasing", nperm = 5000)
    f <- ggplot(subdf, aes(x = subdf[, paste0("tile", m )], y = freq)) + geom_violin()+
      geom_boxplot(outlier.shape = NA, width = 0.1) + geom_jitter(aes(colour = patient)) + 
      scale_color_manual(values  = patient_cols)+
      labs(x = paste("bins of", new_name[m] , "Module mean propression"), y = paste(n, "proportion"))+theme_minimal(base_size = 16)
    
    g <- ggplot(subdf, aes(x = subdf[, paste0("n", m)], y = freq ,colour = patient)) + geom_point() +  
      geom_smooth(method=lm, color="black", se=FALSE)+ 
      scale_color_manual(values =patient_cols)+
      labs(x = paste( new_name[m], "intensity"), y = paste(n, "proportion"))+theme_minimal(base_size = 16)
    cor = cor.test(subdf$freq[!is.na(subdf[, paste0("n", m)])], subdf[, paste0("n", m)][!is.na(subdf[, paste0("n", m)])], method = "spearman")
    print(f+g+plot_annotation(title = paste("correlation between", new_name[m], "Module propression and", n, "proportion:\nrho = ", round(cor$estimate,3), "p-value", round(cor$p.value,3)),
                              subtitle =  paste0("JonckheereTerpstraTest increasing p-value:", test$p.value, " decreasing p-value:",test2$p.value))+
            plot_layout(axes = "collect", guides = "collect"))
    if (any(c(test$p.value, test2$p.value, cor$p.value)<0.05)){
      plotdf <- rbind(plotdf, data.frame(subdf[, 1:7], module = new_name[m], 
                                         prop = subdf[, paste0("n", m)], tile = subdf[, paste0("tile", m )], 
                                         JTincP = test$p.value, 
                                         JTdecP = test2$p.value,
                                         Cor = cor$estimate,
                                         CorP = cor$p.value))
    }
    }else{print("not enough sample")}
  }
  print (n)
}
dev.off()

saveRDS(plotdf, "propModulenormal.Rds")
```
```{r fig.height=4, fig.width=18}
plotdf$module[plotdf$module == "Hypoxia"] <- "Glycolysis"
plotdf$module <- factor(plotdf$module, levels = new_name)

ggplot(plotdf, aes(x = tile, y = freq,fill = module)) + geom_violin()+
  geom_boxplot(outlier.shape = NA, width = 0.1) + 
  geom_jitter(aes(colour = patient), width = 0.1) + 
  scale_color_manual(values  = patient_cols)+
  scale_fill_manual(values = module_cols2)+
  theme_bw()+
  ggh4x::facet_nested_wrap(.~cell_anno+module, scale = "free", nrow = 1)+
  theme(strip.background = element_rect(fill = "white"), 
        axis.title.x = element_text(size = 18),
        strip.text = element_text(size = 16, face = "bold"), 
        axis.title.y = element_text(size = 18),
        legend.position = "bottom", 
        legend.text = element_text(size = 14), 
        legend.title = element_text(size = 16))+
  guides(fill = guide_legend(nrow = 1),
         color = guide_legend(nrow = 1))+
  labs(y = "Proportion", x = "Cell Proportion Levels of Modules", fill = "Module", color = "Patient")
ggsave("pdf/JTtestn.pdf", width = 18, height = 3.5)
```



### correlate module expression level and normal cell proportion
```{r}
normal3 <- readRDS("~/normal_cell_annotation/subtype/final_normalsrt.Rds")
normal3$cell_type <- ifelse(normal3$cell_anno%in% c("T cells","B cells","Plasma cells",  "Macrophages"), "immune cells", 
                            ifelse(normal3$cell_anno%in% c("Fibroblasts","Adipocytes","Pericytes"), "stromal cells", "others"))
normal3$pathology <- ifelse(normal3$patient %in% c("CA0090", "CA0046"), "NE", ifelse(normal3$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))

df <- normal3@meta.data %>% group_by(patient,site,sample, cell_anno) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))

df$pathology <- ifelse(df$patient %in% c("CA0090", "CA0046"), "NE", ifelse(df$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))
df$patient <- gsub("00", "", df$patient)
df$labs <- paste0(signif(df$freq*100, 2), "%")
df$labs[df$freq < 0.03] <- NA

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites
df <- df %>% filter(n >=10) ## Add a cell number filter


df2 <- meta %>% group_by(sample) %>% summarise(avg1 = mean(Module1),
                                                 avg2 = mean(Module2),
                                                 avg3 = mean(Module3),
                                                 avg4 = mean(Module4),
                                                 avg5 = mean(Module5),
                                                 avg6 = mean(Module6)) %>% 
  mutate(tile1 = ordered(ntile(avg1, 5)),
         tile2 =  ordered(ntile(avg2, 5)),
         tile3 =  ordered(ntile(avg3, 5)),
         tile4 =  ordered(ntile(avg4, 5)),
         tile5 =  ordered(ntile(avg4, 5)),
         tile6 =  ordered(ntile(avg6, 5)))

temp  <- merge(df, df2, by = "sample", all.x = T, all.y = F)

plotdf <- data.frame()
pdf("pdf/normalcell_propVSbinedmodule_n=5.pdf", width = 9, height = 5)
for (n in levels(normal3$cell_anno)){
  subdf <-temp[temp$cell_anno == n,]
  for (m in 1:6){
        if (n_distinct(subdf[, paste0("tile", m )])==5 & all(table(subdf[, paste0("tile", m )])>=3)){
    test <- DescTools::JonckheereTerpstraTest(subdf$freq, g = subdf[, paste0("tile", m )], alternative = "increasing", nperm = 5000)
    test2 <- DescTools::JonckheereTerpstraTest(subdf$freq, g = subdf[, paste0("tile", m )], alternative = "decreasing", nperm = 5000)
  f <- ggplot(subdf, aes(x = subdf[, paste0("tile", m )], y = freq)) + geom_violin()+
    geom_boxplot(outlier.shape = NA, width = 0.1) + geom_jitter(aes(colour = patient)) + 
    scale_color_manual(values  = patient_cols)+
    labs(x = paste("bins of", new_name[m] , "Module mean expression"), y = paste(n, "proportion"))+theme_minimal(base_size = 16)
  
  g <- ggplot(subdf, aes(x = subdf[, paste0("avg", m)], y = freq ,colour = patient)) + geom_point() +  
    geom_smooth(method=lm, color="black", se=FALSE)+ 
    scale_color_manual(values =patient_cols)+
    labs(x = paste( new_name[m], "intensity"), y = paste(n, "proportion"))+theme_minimal(base_size = 16)
  cor = cor.test(subdf$freq[!is.na(subdf[, paste0("avg", m)])], subdf[, paste0("avg", m)][!is.na(subdf[, paste0("avg", m)])], method = "spearman")
  print(f+g+plot_annotation(title = paste("correlation between", new_name[m], "Module expression and", n, "proportion:\nrho = ", round(cor$estimate,3), "p-value", round(cor$p.value,3)),
                            subtitle =  paste0("JonckheereTerpstraTest increasing p-value:", test$p.value, " decreasing p-value:",test2$p.value))+
          plot_layout(axes = "collect", guides = "collect"))
  if (any(c(test$p.value, test2$p.value, cor$p.value)<0.05)){
    plotdf <- rbind(plotdf, data.frame(subdf[, 1:7], module = new_name[m], 
                                       exp = subdf[, paste0("avg", m)], tile = subdf[, paste0("tile", m )], 
                                       JTincP = test$p.value, 
                                       JTdecP = test2$p.value,
                                       Cor = cor$estimate,
                                       CorP = cor$p.value))
      
    }
  }else{print("not enough sample")}
  
  }
  print (n)
}
dev.off()
saveRDS(plotdf, "sigModulenormal.Rds")
```

```{r fig.height=4, fig.width=18}
sigModulenormal <- readRDS("~/Fig4-5_archetype/sigModulenormal.Rds")
sigModulenormal%>% select(c("cell_anno", "module")) %>% distinct()
head(sigModulenormal)

sigModulenormal <- readRDS("~/Fig4-5_archetype/sigModulenormal.Rds")
sigModulenormal$module[sigModulenormal$module == "Hypoxia"] <- "Glycolysis"
sigModulenormal$module <- factor(sigModulenormal$module, levels = new_name)
# ggplot(temp1 %>% filter(cell_anno %in% sigModulenormal$cell_anno), aes(x = tile, y = freq,fill = module)) + geom_violin()+
#   geom_boxplot(outlier.shape = NA, width = 0.1) + geom_jitter(aes(colour = patient)) + 
#   scale_color_manual(values  = patient_cols)+
#   scale_fill_manual(values = module_cols2)+
#   theme_minimal(base_size = 16)+
#   facet_wrap(.~cell_anno)
sigModulenormal$cell_anno <- as.character(sigModulenormal$cell_anno)
sigModulenormal$cell_anno[sigModulenormal$cell_anno == "Lymphatic endothelial cells"] <- "Lymphatic\nendothelial"
ggplot(sigModulenormal, aes(x = tile, y = freq,fill = module)) + geom_violin()+
  geom_boxplot(outlier.shape = NA, width = 0.1) + geom_jitter(aes(colour = patient), width = 0.2) + 
  scale_color_manual(values  = patient_cols)+
  scale_fill_manual(values = module_cols2)+
  theme_bw()+
  ggh4x::facet_nested_wrap(.~cell_anno+module, scale = "free", nrow = 1)+
  theme(strip.background = element_rect(fill = "white"), 
        axis.title.x = element_text(size = 18),
        strip.text = element_text(size = 16, face = "bold"), 
        axis.title.y = element_text(size = 18),
        legend.position = "bottom", 
        legend.text = element_text(size = 14), 
        legend.title = element_text(size = 16),
        legend.margin = margin(t = -10),
        strip.clip = "off")+
  guides(fill = guide_legend(nrow = 1),
         color = guide_legend(nrow = 1))+
  labs(y = "Proportion", x = "Expression Levels of Modules", fill = "Module", color = "Patient")
ggsave("pdf/JTtest.pdf", width = 18, height = 3.5)
```

##Hexagon plots
```{r}
library(igraph)
```


```{r}
meta <- readRDS("~/regulon/new_regulon/archetype/final_result/intergrated_modulescore_meta.Rds")
meta$group <- "low expression"
meta$module <- "low expression"

med_cells <- apply(meta[, names(final_signature)], 2, function(x) x > 0.2*(max(x) - min(x)))
ind_med <- apply(med_cells, 1, function(x) if(sum(x) > 1) paste(names(final_signature)[x], collapse = "&"))
for(i in names(final_signature)){
  meta$group[med_cells[, i]] <-""
  meta$module[]
}
meta$group[!sapply(ind_med, is.null)] <- unlist(ind_med)
table(meta$group)


df <- meta %>% group_by(sample.x, group) %>% summarise(n = n()) %>% mutate(freq = n/sum(n))

df$patient <- substr(df$sample.x, 1, 6)
df$group  <- gsub("_", "", df$group)
colnames(df) <- c("sample", "module", "n","freq", "patient")
df$pathology <- ifelse(df$patient %in% c("CA0090", "CA0046"), "NE", ifelse(df$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))


pdf("module_cell_composition_network.pdf", width = 14, height = 10)
par(mfrow= c(7,5), mar = c(1, 1, 1, 1))
for (i in unique(df$sample)){
temp <- df %>% filter(sample == i)
nodes <- temp %>% filter(grepl("positive", module))
nodes <- as.data.frame(nodes)
rownames(nodes) <- gsub(" positive", "",nodes$module)
nodes$id <- rownames(nodes)
nodes <- nodes[, c("id","module", "sample","n", "freq", "patient", "pathology")]
#print(head(nodes))
if(any(!paste("module", 1:6, sep = "_")%in% nodes$id)){
miss_module <- setdiff(paste("module", 1:6, sep = "_"),nodes$id)
nodes <- rbind(nodes, data.frame(id  = miss_module, module = "miss", sample = i, n = 0, freq = 0.00001, patient = unique(nodes$patient), pathology = unique(nodes$pathology)))
}
links <- temp %>% filter(grepl("&", module))
links <- as.data.frame(links)
edges <- apply(links, 1, function(x) {
temp <- strsplit(as.character(x["module"]), split = " ")[[1]][1]
temp <- strsplit(temp, split= "&")[[1]]
temp  <- combn(temp, 2)
return(as.data.frame(t(temp)) %>% `colnames<-`(c("from", "to"))%>% mutate(weight = x["freq"], module = paste(from, to, sep = "&")))
})
edges <- rbindlist(edges)
edges <- edges %>% group_by(module) %>% mutate(weight_sum = sum(as.numeric(weight)))
edges <- edges[, c("from", "to", "weight_sum")] %>% distinct()
#print(head(edges))
net <- graph_from_data_frame(d = edges, vertices = nodes, directed = F)
#net <- simplify(net, remove.multiple = F, remove.loops = T)
V(net)$size <- log(V(net)$freq *100+1)*10
V(net)$color <- colors[names(V(net))]
E(net)$width <- log(E(net)$weight_sum *100 +1)*10
E(net)$edge.color <- apply(edges, 1, function(x) colorRamp2(c(0, 1), colors[c(x[1], x[2])])(0.5))
l <- layout_in_circle(net)
plot(net, layout = l, edge.color = E(net)$edge.color, vertex.label = NA, main  = i)
print(i)
}
dev.off()


```


## sublone distribution
```{r}
subclones <- readRDS("~/Fig3_genomics_normal_cells/ATAClone/subclones.Rds")
for(i in names(subclones)){
  names(subclones[[i]]) <- paste(i,names(subclones[[i]]) , sep = "_")
}
names(subclones) <- NULL
subclones <- unlist(subclones)

meta$subclone <- subclones[rownames(meta)]

meta$group[grep("&", meta$group)] <- "transition"
df <- meta %>% group_by(sample,subclone, group) %>% summarise(n = n()) %>% mutate(freq = n/sum(n))

df$patient <- gsub("00","",substr(df$sample, 1, 6))
df$group  <- gsub(" high", "", df$group)
colnames(df) <- c("sample","subclone", "module", "n","freq", "patient")
df$pathology <- ifelse(df$patient %in% c("CA90", "CA46"), "NE", ifelse(df$patient %in% c("CA27", "CA58"), "Mixed", "AD"))
df$module <- factor(df$module, levels = c(names(final_signature), "transition", "low expression"))
df <- df[!is.na(df$subclone), ]

color_high <- setNames( c("dodgerblue3", brewer.dark2(6)[2:6]),names(final_signature))
colors <- c(color_high, setNames(c("lightgrey", "slategrey"), c("low expression", "transition")))

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites
head(df)
df$x <- paste(df$sample, df$subclone)

lab <- df$site
lab <- paste(lab, sapply(strsplit(df$sample, split = "_"), function(x) tail(x, 1)))
df$lab <- lab
df$lab <- sapply(strsplit(df$sample, split = "_"), function(x) tail(x, 1))
df
saveRDS(df, "subclone_distribution.Rds")
```

```{r fig.height=5, fig.width=20}
ggplot(df, aes(x = x, y = freq, fill =module)) + geom_bar(stat = "identity") + 
  ggh4x::facet_nested(.~patient+lab, scales = "free", space = "free") +
  scale_fill_manual(values = colors, labels = c(new_name, "Transition" ,  "Background")) + 
  theme_bw(base_size = 18)+
  theme(strip.background = element_rect(fill = "white", color = "black"),
        panel.spacing = unit(0.05, "cm"), 
        strip.text.x.top = element_text(size = 20, face = "bold"), 
        legend.margin = margin(0, 0, 0, -10), 
        axis.text.y.left = element_text(angle = 90, hjust = 0.5))+
  scale_x_discrete(breaks = df$x, labels= df$subclone)+
  labs(x = "Subclones", y = "Cell Proportion", fill = "Module")
```
```{r}
ggsave("pdf/subclones.pdf", width = 18, height = 3.5)
```

