---
title: "R Notebook"
output: html_notebook
---



```{r set up, include=FALSE}
library(ggplot2)
library(pals)
library(scales)
library(Seurat)
library(colorRamps)
library(ggpubr)
library(geomtextpath)
library(ggpubr)
library(dplyr)
library(circlize)
library(patchwork)
library(data.table)
library(stringr)


meta <- readRDS("~/CASCADEpaper/paper/Fig5_archetype/5_integrated_final_module_meta.Rds")
final_signature <- readRDS("~/CASCADEpaper/paper/Fig5_archetype/clean_module.Rds")
# srt <- readRDS("~/integration/2024_06/tumor_normal_ruvsrt.Rds")
new_name <- c("AR","Inflammation", "NE1","NE2", "Cycling","Glycolysis"  )
names(final_signature) <- paste0("Module", 1:6)
phenotype_meta <- readRDS("~/CASCADEpaper/paper/PSMA/phenotype_meta.Rds")



source("~/CASCADEpaper/paper/functions.R")

```




## normal tumor expression 
```{r}
srt <- readRDS("~/integration/2024_06/tumor_normal_ruvsrt.Rds")
srt$cell_anno <- gsub("00", "", srt$cell_anno)
srt$cell_anno <- str_wrap(srt$cell_anno, 20)
```


```{r}
f <- FeaturePlot(srt, raster = F, "FOLH1")
coneraxes(f)
ggsave("FOLH1exp.pdf", height = 5, width = 6)
```

## noraml expression 
```{r}
# pdf("FOLH1exp_normal.pdf", height = 5, width = 7)
# VlnPlot(srt, "FOLH1", group.by = "cell_anno", cols = c(brewer.set1(14), hue_pal()(9)))
# dev.off()

srt$FOLH1exp <- srt[["ruv3"]]@data["FOLH1", ]
saveRDS(srt@meta.data, "tumor_normal_meta.Rds")
tumor_normal_meta <- readRDS("~/CASCADEpaper/paper/PSMA/tumor_normal_meta.Rds")
f <- ggplot(tumor_normal_meta %>% filter(!grepl("CA", tumor_normal_meta$cell_anno)& FOLH1exp > 0), aes(x = cell_anno, y = FOLH1exp, fill = cell_anno))+theme_bw(base_size = 18)+
  # geom_violin()+
  geom_boxplot(color="black")+
  # geom_jitter(width = 0.1, alpha = 0.1)+
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1), legend.position = "none", 
        legend.margin = margin(r= 20, l = 20, t = -20), legend.title.position = "top")+
  scale_fill_manual(values =  brewer.set1(14), name = "normal cells")+
  guides(fill = guide_legend(nrow = 3)) + xlab("") + ylab("FOLH1 expression")
f
ggsave("FOLH1_normalbox.pdf", width = 6, height = 4.5)

```
## correlation with normal cell proportion
```{r}
normal3 <- readRDS("~/CASCADEpaper/paper/normal_cells_202406/subtype/final_normalsrt.Rds")
normal3$cell_type <- ifelse(normal3$cell_anno%in% c("T cells","B cells","Plasma cells",  "Macrophages"), "immune cells", 
                            ifelse(normal3$cell_anno%in% c("Fibroblasts","Adipocytes","Pericytes"), "stromal cells", "others"))
normal3$pathology <- ifelse(normal3$patient %in% c("CA0090", "CA0046"), "NE", ifelse(normal3$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))

df <- normal3@meta.data %>% group_by(patient,site,sample, cell_anno) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))

df$pathology <- ifelse(df$patient %in% c("CA0090", "CA0046"), "NE", ifelse(df$patient %in% c("CA0027", "CA0058"), "Mixed", "AD"))
df$patient <- gsub("00", "", df$patient)
df$labs <- paste0(signif(df$freq*100, 2), "%")
df$labs[df$freq < 0.03] <- NA

sites <- df$sample
sites[grep("brain", sites)] <- "brain"
sites[grep("dura", sites)] <- "dura"
sites[grep("prostate", sites)] <- "prostate"
sites[grep("liver", sites)] <- "liver"
sites[grep("fat", sites)] <- "fat"
sites[grep("lymph|hilar", sites)] <- "LN"
sites[grep("rib|vertebra", sites)] <- "bone"
sites[grep("lung", sites)] <- "lung"
sites[grep("abdomen", sites)] <- "abdomen"
sites[grep("bladder", sites)] <- "bladder"
df$site <- sites

df2 <- phenotype_meta %>% group_by(sample) %>% summarise(gene = "FOLH1", exp = mean(FOLH1_exp)) %>% mutate(tile = ordered(ntile(exp, 5)))

temp  <- merge(df, df2, by = "sample", all.x = T, all.y = F)

plotdf <- data.frame()
pdf("normalcell_propVSbineFOLH1_n=5.pdf", width = 10, height = 5)
for (n in levels(normal3$cell_anno)){
  subdf <-temp[temp$cell_anno == n,]
    test <- DescTools::JonckheereTerpstraTest(subdf$freq, g = subdf$tile, alternative = "increasing", nperm = 5000)
    test2 <- DescTools::JonckheereTerpstraTest(subdf$freq, g = subdf$tile, alternative = "decreasing", nperm = 5000)

  f <- ggplot(subdf, aes(x =tile, y = freq)) + geom_violin()+
    geom_boxplot(outlier.shape = NA, width = 0.1) + geom_jitter(aes(colour = patient)) + 
    scale_color_manual(values  = patient_cols)+
    labs(x = paste("bins of FOLH1 \n average expression"), y = paste(n, "proportion"))+theme_minimal(base_size = 16)
  
  g <- ggplot(subdf, aes(x = subdf$exp, y = freq ,colour = patient)) + geom_point() +  
    geom_smooth(method=lm, color="black", se=FALSE)+ 
    scale_color_manual(values =patient_cols)+
    labs(x = paste("FOLH1 intensity"), y = paste(n, "proportion"))+theme_minimal(base_size = 16)
  cor = cor.test(subdf$freq[!is.na(subdf$exp)], subdf$exp[!is.na(subdf$exp)], method = "spearman")
  print(f+g+plot_annotation(title = paste("correlation between FOLH1 expression and", n, "proportion:\nrho = ", round(cor$estimate,3), "p-value", round(cor$p.value,3)),
                            subtitle =  paste0("JonckheereTerpstraTest increasing p-value:", test$p.value, " decreasing p-value:",test2$p.value))+
          plot_layout(axes = "collect", guides = "collect"))
  if (any(c(test$p.value, test2$p.value, cor$p.value)<0.05)){
    plotdf <- rbind(plotdf, data.frame(subdf[, 1:7],
                                       exp = subdf$exp, tile = subdf$tile, 
                                       JTincP = test$p.value, 
                                       JTdecP = test2$p.value,
                                       Cor = cor$estimate,
                                       CorP = cor$p.value))
  }
  print (n)
}
dev.off()

saveRDS(plotdf, "FOLH1normal.Rds")
```


```{r}
plotdf$sigP <- apply(plotdf[, c("JTdecP", "JTincP")], 1, function(x) x[x<0.05])

plotdf %>% group_by(cell_anno) %>% summarise(x  = 1, xtile = 3, y = 0.7, labs = paste("JT p value =", unique(sigP)), rhoL = paste("rho =", signif(unique(Cor), 2), "p=", signif(unique(CorP), 2))) -> labdf


g1 <- ggplot(plotdf, aes(x = tile, y = freq)) + geom_violin()+
  geom_boxplot(outlier.shape = NA, width = 0.1) + geom_jitter(aes(colour = patient)) + 
  theme_minimal(base_size = 16)+
  #geom_text(data = labdf, aes(label = labs, x = xtile, y = y))+
  facet_grid(cell_anno~., scales = "free_y")+
  labs(x = "FOLH1 Expression Tile", y = "Cell Proportion")+
  theme(legend.margin = margin(-10, 0, 0, 0), 
        strip.text = element_text(face = "bold"))

g2 <- ggplot(plotdf, aes(x = exp, y = freq)) + geom_point(aes(colour = patient)) +  
    geom_smooth(method=lm, color="black", se=FALSE)+ 
  #geom_text(data = labdf, aes(label = rhoL, x = x, y = y), hjust = 0.5)+
    # scale_color_manual(values =patient_cols)+
    labs(x = paste("FOLH1 Intensity"), y = "Cell Proportion")+theme_minimal(base_size = 16)+
    facet_grid(cell_anno~., scales = "free")+
  theme(legend.margin = margin(-10, 0, 0, 0), 
        strip.text = element_text(face = "bold"))

g1 +g2 + plot_layout(axes = "collect", guides = "collect") & theme(legend.position = "bottom")
ggsave("normalJTtest.pdf", width = 5, height = 4)

```





##tumor only expression
```{r}
srt <- readRDS("~/integration/2024_06/tumor_only_ruvsrt.Rds")
f <- FeaturePlot(srt, raster = F, "FOLH1")
f
FeaturePlot(srt, raster = F, "AR")
FeaturePlot(srt, raster = F, "ASCL1")
FeaturePlot(srt, raster = F, "DLL3")
FeaturePlot(srt, raster = F, "STEAP1")
FeaturePlot(srt, raster = F, "STEAP2")
FeaturePlot(srt, raster = F, "CD276") # B7H3
FeaturePlot(srt, raster = F, "HOXB13")
FeaturePlot(srt, raster = F, "ROR1")


pdf("tumouronly_featurescatter.pdf", width = 7, height = 4)
p1 <- FeatureScatter(srt, "FOLH1", "AR")
p2 <- FeatureScatter(srt, "FOLH1", "ASCL1")
p1 + p2 
dev.off()

ggplot(phenotype_meta, aes(x = FOLH1_exp, y = AR_exp)) + geom_point()
cor.test(phenotype_meta$FOLH1_exp, phenotype_meta$AR_exp, method = "spearman")
```

```{r}
pdf("all_markers_tumouronly.pdf", width = 12, height  = 12)
FeaturePlot(srt, c("FOLH1","ASCL1","AR",  "STEAP1", "STEAP2", "ROR1", "DLL3", "CD276", "HOXB13"))
dev.off()
```

## NE expression
```{r fig.height=5, fig.width=5}

f <- FeaturePlot(srt, "FOLH1", cells = colnames(srt)[grep("CA0090|CA0046|CA0058_liver", srt$sample)])
f <- coneraxes(f)
ggsave("FOLH1exp_in_NE.pdf",width = 5, height = 5)

```


```{r}
srt$FOLH1_exp <- srt@assays$ruv3@data["FOLH1", ]
srt$AR_exp <- srt@assays$ruv3@data["AR", ]
srt$HOXB13 <- srt@assays$ruv3@data["HOXB13", ]
srt$ASCL1_exp<- srt@assays$ruv3@data["ASCL1", ]
srt$DLL3_exp<- srt@assays$ruv3@data["DLL3", ]
srt$STEAP2_exp <- srt@assays$ruv3@data["STEAP2", ]
srt$STEAP1_exp <- srt@assays$ruv3@data["STEAP1", ]
srt$CD276_exp <- srt@assays$ruv3@data["CD276", ]
srt$ROR1_exp <- srt@assays$ruv3@data["ROR1", ]
```

## phenotype: high/low
```{r}
pheno <- apply(srt@meta.data[, grep("_exp", colnames(srt@meta.data))], 2, function(x){
  bins <- cut(x, breaks = 10)
  pheno <- ifelse(x== 0, "neg",ifelse(bins %in% levels(bins)[5:10] , "high", "low"))
  return(pheno)
})
head(pheno)

pheno <- apply(pheno, 1, function(x) paste0(c("FOLH1", "AR", "ASCL1", "STEAP2", "STEAP1", "DDL3", "CD276"), x))
pheno <- as.data.frame(t(pheno))
colnames(pheno) <- c("FOLH1_pheno", "AR_pheno", "ASCL1_pheno", "STEAP2_pheno", "STEAP1_pheno", "DDL3_pheno", "CD276_pheno")
srt <- AddMetaData(srt, pheno)
srt <- AddMetaData(srt, meta)
srt$patient <- gsub("00", "", srt$patient)
srt$pheno <- paste0(srt$FOLH1_pheno, srt$AR_pheno)
srt$group <- ifelse(grepl("CA0090|CA0046|CA0058_liver", srt$sample), "AR-/NE+", 
                        ifelse(grepl("CA0027|CA0058_hilar", srt$sample), "AR+/NE+", "AR+/NE-"))
srt$group <- factor(srt$group, levels = c("AR+/NE-", "AR+/NE+", "AR-/NE+"))
table(srt$pheno)
saveRDS(srt@meta.data, "phenotype_meta.Rds")
```

```{r}
library(ggh4x)
site = structure(brewer.set3(10), names = unique(phenotype_meta$site))

f <- ggplot(phenotype_meta, aes(x = sample, y = FOLH1_exp, fill = site))+theme_bw(base_size = 16)+
  geom_boxplot(color="black", outlier.size = 0.2)+
  scale_fill_manual(values = site)+
  # geom_jitter(width = 0.1, alpha = 0.1)+
  facet_nested(.~group+patient, space = "free", scales = "free")+ 
  theme(axis.text.x = element_text(angle =90, vjust = 0.5, hjust = 1), panel.spacing = unit(0,"line"),
        legend.margin = margin(t= -10),
        strip.background = element_rect(fill = "white"), 
        legend.position = "right", 
        axis.title.x = element_blank())+
  scale_x_discrete(labels= xlab) + labs(y = "FOLH1 Expression", fill = "Site")
f

ggsave("FOLH1_box.pdf", width = 12.5, height = 4)
```
```{r}
ggplot(phenotype_meta, aes(x = site, y = FOLH1_exp, fill = site)) + 
  geom_boxplot()+
  scale_fill_manual(values = site)+
  theme_minimal(base_size = 18)+
  theme(legend.position = "none",axis.text.x = element_text(angle =90, vjust = 0.5, hjust = 1))+
  labs(y = "FOLH1 Expression", x = "Site")

ggsave("FOLH1_site_box.pdf", width = 5, height = 4)
```


```{r}
f <- FeaturePlot(srt, "FOLH1")
coneraxes(f)
ggsave("FOLH1exp_tumoronly.pdf",width = 5, height = 5)
```

```{r}
df <- phenotype_meta%>% filter(group == "AR-/NE+") %>% group_by(patient,site,group,sample, FOLH1_pheno) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))

f <- ggplot(df, aes(x = sample, y = freq, fill = FOLH1_pheno)) + geom_bar(stat = "identity") + 
  facet_grid(.~patient, scales = "free", space = "free") + 
   theme_minimal(base_size = 18) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), legend.position = "bottom", 
        legend.margin = margin(-15, 0, 0, -50), 
        axis.title.x = element_blank(), 
        legend.title = element_text(size = 16))+
  scale_fill_manual(values =c("#31A354", "steelblue"), name = "FOLH1\nphenotype")+
  scale_x_discrete(labels= xlab)+
  labs(y = "Cell Proportion")
f
ggsave("FOLH1_bar_NE.pdf", width = 4, height = 4)


```
##phenotype in adeno
```{r}
df <- phenotype_meta %>% filter(group %in% c( "AR+/NE-", "AR+/NE+") )%>% group_by(patient,site,group,sample, pheno ) %>% summarise(n = n()) %>% mutate (freq = n/sum(n))
df$patient <- gsub("CA", "", df$patient)
f <- ggplot(df, aes(x = sample, y = freq, fill = pheno)) + geom_bar(stat = "identity") + 
  ggh4x::facet_nested(.~group+patient, scales = "free", space = "free") + 
   theme_classic(base_size = 18) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 13),legend.margin = margin(t = 0, r = -10, l = -10), 
        axis.title.x = element_blank(), 
        legend.text = element_text(size = 14), 
        panel.spacing = unit(3, "pt"))+
  scale_fill_manual(values = c(rev(brewer.oranges(3)),rev(brewer.greens(3)), rev(brewer.blues(3))) , name = "phenotype")+
  scale_x_discrete(labels= xlab)+
  labs(y = "Cell Proportion")
f
ggsave("FOLH1_bar.pdf", width = 7, height = 4)

```
```{r}
f <- DimPlot(srt, group.by = "pheno", cells = colnames(srt)[srt$group %in% c( "AR+/NE-", "AR+/NE+") ], 
        cols = c(rev(brewer.oranges(3)),rev(brewer.greens(3)), rev(brewer.blues(3))))+NoLegend()
coneraxes(f)
ggsave("pheno_umap.pdf", width = 5, height = 5)
```

###CA27 integrated expression
```{r}
CA0027_ruvsrt <- readRDS("~/integration/2024_06/perpatient/tumor_only/CA0027_ruvsrt.Rds")
f <- FeaturePlot(CA0027_ruvsrt, c("FOLH1", "ASCL1"), ncol = 1)
df <- as.data.frame(t(CA0027_ruvsrt[["RNA2"]][c("FOLH1", "ASCL1"), ]))
df <- df[df$ASCL1>0,]
f2 <- ggplot(df, aes(x = FOLH1, y = ASCL1)) +geom_point() + theme_minimal(base_size = 20)

f|f2 + plot_annotation(title = "CA27 ASCL1 and FOLH1 expression")
ggsave("CA27 ASCL1 and FOLH1 expression.pdf", width = 10, height = 6)

```


## scatter plot between modules signature and gene expression 
```{r}
meta = phenotype_meta <- readRDS("~/CASCADEpaper/paper/PSMA/phenotype_meta.Rds")
cells <- apply(meta[, grep("Module", colnames(meta))], 2, function(x){
  bins <- cut(x, breaks = 10)
  group <- bins %in% levels(bins)[5:10]
  return(setNames(group, rownames(meta)))
})

colnames(cells) <- paste0(colnames(cells), "_group")
phenotype_meta <- cbind(phenotype_meta, cells)
phenotype_meta
average_expression <- function(gene){

  exp <- apply(cells, 2, function(x) mean(meta[x, paste0(gene, "_exp")]))
  return(exp)
}
```

```{r}
df <- data.frame(AR = average_expression("AR"), FOLH1 = average_expression("FOLH1"), 
                 STEAP1 = average_expression("STEAP1"), STEAP2 = average_expression("STEAP2"),
                 ROR1 = average_expression("ROR1"), CD276 = average_expression("CD276"))
df$module <- factor(new_name, new_name)
```


```{r}
f1 <- ggplot(df, aes(x = AR, y = FOLH1, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))


f2 <- ggplot(df, aes(x = AR, y = STEAP1, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))

f3 <- ggplot(df, aes(x = AR, y = ROR1, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))
f7 <- ggplot(df, aes(x = AR, y = CD276, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))


f4 <- ggplot(df, aes(x = FOLH1, y = ROR1, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))


f5 <- ggplot(df, aes(x = FOLH1, y = STEAP1, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))

f6 <- ggplot(df, aes(x = FOLH1, y = STEAP2, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))

f8 <- ggplot(df, aes(x = FOLH1, y = CD276, color = module))+geom_point() +theme_bw(base_size = 16)+
  scale_color_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))



f1+f2+f3+f7+f4+f5+f6+f8+plot_layout(ncol = 4, nrow = 2, guides = "collect")&labs(color = "Module")&theme(legend.margin = margin(-10, 0, 0, 0), legend.position = "bottom")

```
```{r}
pdf("module_scatter.pdf", width = 9, height = 4)
f1+f2+f3+f7+f4+f5+f6+f8+plot_layout(ncol = 4, nrow = 2, guides = "collect")&labs(color = "Module")&theme(legend.margin = margin(-10, 0, 0, 0), legend.position = "bottom")&guides(color = guide_legend(nrow = 1))
dev.off()
```

```{r}
extract_expression <- function(gene){
  cells <- apply(phenotype_meta[, grep("_group", colnames(phenotype_meta))],2, function(x){
    rownames(phenotype_meta)[x]
  })
  names(cells) <- new_name
  exp <- sapply(cells, function(x) return(phenotype_meta[x, paste0(gene, "_exp")]))
  exp <- lapply(new_name, function(x) data.frame(module = x, expression = exp[[x]]))
  return(rbindlist(exp))
}

df2 <- lapply(c("FOLH1", "AR", "ASCL1", "ROR1", "STEAP1", "STEAP2", "CD276"), extract_expression)
names(df2) <- c("FOLH1", "AR", "ASCL1", "ROR1", "STEAP1", "STEAP2", "CD276")



```

```{r}
f <- ggplot(rbindlist(df2, idcol = "gene"), aes(x = module, y = expression, fill = gene)) + 
    geom_boxplot()+
   theme_bw(base_size = 18) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),legend.margin = margin(t = 0, r = -10, l = -10))
  # scale_fill_manual(values = c(rev(brewer.oranges(3)),rev(brewer.greens(3)), rev(brewer.blues(3))) , name = "phenotype")+
  # scale_x_discrete(labels= xlab)
f
ggsave("all_markers_moudle_boxplot.pdf", width = 6, height = 4)
df2 <- rbindlist(df2, idcol = "gene") %>% filter(gene == "FOLH1")

df2$module <- factor(df2$module, levels = as.character(rev(df2 %>% group_by(module) %>% summarise(mean = mean(expression)) %>% arrange(mean) %>% pull(module))))
compare_means(expression~module, data = df2)
ggplot(df2, aes(x = module, y = expression, fill = module)) + 
    geom_boxplot()+
   theme_minimal(base_size = 18) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),legend.margin = margin(t = 0, r = -10, l = -10), 
        axis.title.x = element_blank())+
  scale_fill_manual(values = module_cols2)+
  labs(y = "FOLH1 expression", fill = "Module")
ggsave("FOLH1_moudle_boxplot.pdf", width = 4, height = 4)
```
```{r}
head(phenotype_meta)
phenotype_meta$pheno %>% unique()
mat2 <- phenotype_meta[, c("pheno", paste0("Module", 1:6))]
colnames(mat2) <-  c("pheno", new_name)
df3 <- reshape2::melt(mat2 )
colnames(df3) <- c("pheno", "module", "expression")
df3 <- df3 %>% filter(module %in% c("AR", "Inflammation", "Glycolysis"))
df3$group <- ifelse(grepl("ARhigh", fixed = T, df3$pheno), "ARhigh", ifelse(grepl("ARlow", fixed = T, df3$pheno), "ARlow", "ARneg"))
p2 <- ggplot(df3, aes(x = module, y = expression, fill =pheno)) + 
    geom_boxplot(outliers = T)+
   theme_bw(base_size = 18) + 
  scale_fill_manual( values = c(rev(brewer.oranges(3)),rev(brewer.greens(3)), rev(brewer.blues(3))) , name = "phenotype")+
  theme(axis.text.x = element_blank(),
        axis.ticks.length.x = unit(0, "cm"),
        axis.title.x = element_blank(), 
        legend.margin = margin(t = 0, r = -10, l = -10),
        panel.spacing = unit(0,"line"),
        strip.background = element_rect(fill = "white"), 
        legend.position = "none")+ylab("module")+
  ggh4x::facet_nested(group ~ module, space = "free", scale = "free")+
  labs(y = "module expression")
  
p2
ggsave("FOLH1_moduleexp.pdf", width = 8, height = 4)
```

```{r}
df4 <- df3 %>% mutate(FOLH1_exp = rep(phenotype_meta$FOLH1_exp, 3)) %>% group_by(module, group) %>% mutate(correlation = signif(cor(FOLH1_exp, expression, method = "spearman"), 2), corP = signif(cor.test(FOLH1_exp, expression, method = "spearman")$p.value, 2))
df4 %>% group_by(module, group) %>% summarise(correlation = signif(cor(FOLH1_exp, expression, method = "spearman"), 2), corP = signif(p.adjust(cor.test(FOLH1_exp, expression, method = "spearman")$p.value, method = "BH"), 2))

ggplot(df4, aes(x = FOLH1_exp, y = expression, color =pheno)) + 
    geom_point()+
  geom_labelsmooth(aes(label = correlation), fill = "white",
                method = "lm", formula = y ~ x,
                size = 4, linewidth = 1, boxlinewidth = 0.4, color = "black")+
   theme_bw(base_size = 18) + 
  scale_color_manual( values = c(rev(brewer.oranges(3)),rev(brewer.greens(3)), rev(brewer.blues(3))) , name = "phenotype")+
  theme(legend.margin = margin(t = 0, r = -10, l = -10),
        panel.spacing = unit(0,"line"),
        strip.background = element_rect(fill = "white"), 
        legend.position = "none")+ylab("module")+
  ggh4x::facet_nested(group ~ module, space = "free", scale = "free")+
  labs(y = "Module Expression", x = "FOLH1 Expression")
ggsave("FOLH1_moduleexp_correlation.pdf", width = 5, height = 4)
```
```{r}
df3 %>% mutate(FOLH1_exp = rep(phenotype_meta$FOLH1_exp, 3)) %>% group_by(module, group) %>% summarise(correlation = signif(cor(FOLH1_exp, expression, method = "spearman"), 2), corP = signif(cor.test(FOLH1_exp, expression, method = "spearman")$p.value, 2))
```


```{r}
df2 <- extract_expression("FOLH1")
df2 <- df2 %>% filter(module %in% c("AR", "Inflammation", "Glycolysis"))
df2$module <- factor(df2$module, levels = new_name)
p1 <- ggplot(df2, aes(x = module, y = expression, fill = module)) + 
    # geom_violin()+
    geom_boxplot(width = 0.8)+
   theme_bw(base_size = 18) + 
  theme(legend.margin = margin(t = 0, r = -10, l = -10), legend.position = "none")+ylab("FOLH1")+
  scale_fill_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))+
  xlab("") + ylab("FOLH1")
p1
```


```{r}
library(ggpubr)
compare_means(expression~module, df2, method = "wilcox.test")
ggboxplot(df2, x = "module", y = "expression") +  stat_compare_means(comparisons =combn(unique(as.character(df2$module)), 2, simplify = F), method = "wilcox")

```
```{r}
p1 + p2 + plot_layout(nrow = 2, axes = "collect", guides = "collect")

```

```{r}
lapply(names(df2), function(temp2){
  temp <- df2[[temp2]]
  ggplot(temp, aes(x = module, y = expression, fill = module)) + 
    # geom_violin()+
    geom_boxplot(width = 0.8)+
   theme_bw(base_size = 18) + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),legend.margin = margin(t = 0, r = -10, l = -10), legend.position = "none")+ggtitle(temp2)+
  scale_fill_manual(values = c("dodgerblue3", brewer.dark2(6)[2:6]))
})
```





## DEG and gene set enrichment

```{r}
library(MAST)
com1 <- combn(grep("ARhigh", unique(srt$pheno), value = T), 2, simplify = F)
x <- com1[[1]]
Idents(srt) <- srt$group
deg <- lapply(com1, function(x){
  marker <- FindMarkers(srt, ident.1 =  x[1], ident.2 = x[2], test.use = "MAST", 
                      min.cells.feature = 20,min.cells.group = 20, group.by = "pheno", subset.ident = c("AR+/NE-", "AR+/NE+"))
})
```
```{r}
library(patchwork)
library(dplyr)
library(Seurat)
library(data.table)
library(pals)
library(msigdbr)
library(clusterProfiler)
library(enrichplot)
library(ggnewscale)
library(dplyr)
library(tibble)
library(data.table)
library(ggstance)
library(MAST)
library(gridExtra)
library(stringr)
library(ggplot2)
library(tidyverse)


Hall <- msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gs_name, gene_symbol)
KEGG <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "KEGG")%>%
  dplyr::select(gs_name, gene_symbol)
Reactome <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "REACTOME")%>%
  dplyr::select(gs_name, gene_symbol)

C5 <- msigdbr(species = "Homo sapiens", category = "C5") %>%
  dplyr::select(gs_name, gene_symbol)
C6 <- msigdbr(species = "Homo sapiens", category = "C6") %>%
  dplyr::select(gs_name, gene_symbol)
C7 <- msigdbr(species = "Homo sapiens", category = "C7") %>%
  dplyr::select(gs_name, gene_symbol)
C8 <- msigdbr(species = "Homo sapiens", category = "C8") %>%
  dplyr::select(gs_name, gene_symbol)

enrichment_barplots <- function(geneList, geneSets, groupname){
  enresult <- try(GSEA(geneList, TERM2GENE = geneSets))
  if (class(enresult)!= "try-error"){
    if(nrow(enresult@result)>0 ){
      y <- arrange(enresult, abs(NES)) %>% 
        group_by(sign(NES)) %>% 
        mutate(Description = gsub("_", " ", Description)) %>%
        mutate(Description = str_wrap(Description, width = 20)) %>% 
        dplyr::slice(1:5)
      f <- ggplot(y, aes(NES, fct_reorder(Description, NES), fill=p.adjust), showCategory=10) + 
        geom_col(orientation='y') + 
        scale_fill_continuous(low='blue', high='red', guide=guide_colorbar(reverse=TRUE)) + 
        theme_minimal() + ylab(NULL) + ggtitle(groupname)
      print(f)
      return(enresult)}
    else{
      print("no enrichment")
    }
  }else{print("not enough gene mapped")}
}

generate_genelist <- function(markerdf){
  markerdf <- markerdf[markerdf$p_val_adj <0.05, ]
  temp <- setNames(markerdf$avg_log2FC, rownames(markerdf))
  temp <- sort(temp, decreasing = T)
  return(temp)
}
```

```{r}
ARhighDEG <- readRDS("~/CASCADEpaper/paper/PSMA/ARhighDEG.Rds")
ARlowDEG <- readRDS("~/CASCADEpaper/paper/PSMA/ARlowDEG.Rds")
ARnegDEG <- readRDS("~/CASCADEpaper/paper/PSMA/ARnegDEG.Rds")
```

```{r}
lapply(ARlowDEG, function(marker){
  lapply(list(Hall, KEGG, Reactome, C5, C6), function(g){
    enrichment_barplots(generate_genelist(marker), g, paste(unique(marker$group1), "VS", unique(marker$group2)))
  })
})
```

```{r}
lapply(ARnegDEG, function(marker){
  lapply(list(Hall, KEGG, Reactome, C5, C6), function(g){
    enrichment_barplots(generate_genelist(marker), g, paste(unique(marker$group1), "VS", unique(marker$group2)))
  })
})
```

## CN of PSMA
```{r}
geneCNAlist <- readRDS("~/CASCADEpaper/paper/Fig3_genomics/copy_number/geneCNAlist.Rds")
str(geneCNAlist)
names(geneCNAlist)[c(8,9)] <- c("CA0035_bladder_2" , "CA0035_paraaortic_lymph_node_1")
cns <- lapply(geneCNAlist, function(x) x %>% filter(gene == "FOLH1"))
cns <- rbindlist(cns, idcol = "sample", fill = T)
head(cns)
head(phenotype_meta)


meta <- phenotype_meta[, c("sample","patient","site")]
meta$pathology <- ifelse(meta$patient %in% c("CA90", "CA46"), "NE", ifelse(meta$patient %in% c("CA27","CA58"), "Mixed", "AD"))
meta <- distinct(meta)
cns <- merge(cns, meta, by = "sample")
df <- phenotype_meta %>% group_by(sample) %>% summarise(meanexp = mean(FOLH1_exp))
cns <- merge(cns, df, by = "sample")
saveRDS(cns, "FOLH1_CNV.Rds")

xlab <- paste(cns$patient, cns$site, sapply(strsplit(cns$sample, split = "_"), function(x) tail(x, 1)))#, sapply(strsplit(cns$sample, split = "_"), function(x) tail(x, 1)))
xlab <- setNames(xlab, cns$sample)

ggplot(cns, aes(x = sample, y = CN, fill = patient))+
  theme_classic(base_size = 16)+
  geom_bar(stat = "identity")+
  scale_x_discrete(labels= xlab)+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
        legend.margin = margin(t = 20, l = -10), 
        panel.spacing = unit(3, "pt"), 
        axis.title.x = element_blank(), 
        axis.title.y = element_text(size = 14))+
  facet_grid(.~pathology, scale = "free", space = "free")+
  geom_abline(slope = 0, intercept = 2)+
  geom_point(aes(y = meanexp/max(cns$meanexp)*7))+
  scale_y_continuous(
    name = "CN",                 # Primary y-axis label
    sec.axis = sec_axis(~ . /7*max(cns$meanexp), name = "Average\nFOLH1") # Secondary y-axis label and rescale factor
  )+
  guides(fill= guide_legend(ncol = 1))+
  labs(fill = "Patient")
  
ggsave("FOLH_CN.pdf", width =6.5, height = 4)


```
## subclone DE of PSMA
```{r}
DEGs <- readRDS("~/CASCADEpaper/paper/Fig3_genomics/ATAClone/DEGs.Rds")
DEGs <- lapply(DEGs, function(x) rbindlist(x))
test <- lapply(DEGs, function(x) x[x$gene == "FOLH1" &x$p_val_adj < 0.05, ])
test <- rbindlist(test, idcol = "sample", fill = T)
test$id <- paste(test$sample, test$comp)

DE_region <- readRDS("~/CASCADEpaper/paper/Fig3_genomics/ATAClone/DE_region.Rds")
test2  <- lapply(DE_region, function(x){
  x[["region"]] %>% filter(gene == "chr11.p:42720000-53400000") %>% filter(p_val_adj < 0.05)
 })
test2 <- rbindlist(test2, fill = T, idcol = "sample")
test2$id <- paste(test2$sample, test2$comp)

head(test)
head(test2)

temp <- merge(test, test2, by = "id", all = T)

```

```{r }
library(scran)
library(scater)
library(ComplexHeatmap)

paths <- system("realpath /trigos_team/CASCADE/Analysis/230113_seurat_intron/*/*/*non_tumour_removed.Rds", intern = T)
paths_split <- unlist(lapply(strsplit(paths, split = "/"), function(x) paste(x[6], x[7], sep = "_")))
paths2 <- system("realpath /trigos_team/CASCADE/Analysis/240603_seurat_intron/CA0027/dura*/*non_tumour_removed_v4.Rds", intern = T)
paths_split2 <- unlist(lapply(strsplit(paths2, split = "/"), function(x) paste(x[6], x[7], sep = "_")))

paths <- c(paths, paths2)
paths_split <- c(paths_split, paths_split2)
paths <- setNames(paths,paths_split)
```


```{r CA0035 bladder }
i = "CA0035_bladder_2"
patient = "CA0035"
sample = "bladder_2"
srt<- readRDS(paths[i])
  
  DefaultAssay(srt) <- "RNA"
  lib.median <- median(srt$nCount_RNA)
  target_pseudocount <- 1
  srt <- srt %>% 
    NormalizeData(scale.factor = lib.median / target_pseudocount) %>% 
    ScaleData(do.scale = F)
  
  set.seed(100)
  poisson_fit2 <- modelGeneVarByPoisson(srt@assays$RNA@counts, dispersion = 0.2, size.factors = target_pseudocount * srt$nCount_RNA / lib.median)
  residuals2 <- poisson_fit2$total - poisson_fit2@metadata$trend(poisson_fit2$mean)
  names(residuals2) <- rownames(srt)
  top_genes2 <- rownames(srt)[order(residuals2, decreasing = TRUE)]
  
  srt <- srt %>% 
    RunPCA(features = top_genes2[1:2000]) %>% 
    FindNeighbors(dims = 1:20) %>% 
    FindClusters(resolution = 1) %>% 
    RunUMAP(dims = 1:20)
  
  joint_cn <- readRDS(paste("/trigos_team/CASCADE/Analysis/240628_ATAClone", patient, sample, "joint_cn.Rds", sep = "/"))
  subclone <- readRDS(paste("/trigos_team/CASCADE/Analysis/240628_ATAClone", patient, sample, "leiden_cluster.Rds", sep = "/"))
  srt$subclone <- subclone[colnames(srt)]
  Idents(srt) <- srt$subclone
  ind <- setdiff(levels(srt$subclone), names(which.max(sapply(levels(srt$subclone), function(x) sum(joint_cn[, colnames(srt)[which(srt$subclone == x)]] %in% c(2, 1))))))
  
  g <- DimPlot(srt, group.by = "subclone", cells =colnames(srt)[srt$subclone %in% ind])
  g <- DimPlot(srt, group.by = "subclone", cells =colnames(srt)[srt$subclone %in% ind])
  g <- coneraxes(g)
  g2 <- VlnPlot(srt, features = c("FOLH1"), group.by = "subclone", idents = ind)
  g|g2
  ggsave("CA35_bladder_subclone.pdf",g|g2, width = 6, height = 4)
  temp <- joint_cn["chr11.p:42720000-53400000", ]
  srt$subcloneCN <- temp[colnames(srt)]
  FeaturePlot(srt, "subcloneCN")
  cnmat <- matrix(joint_cn["chr11.p:42720000-53400000", colnames(srt)[srt$subclone %in% ind]], nrow = 1)
  rols <- as.character(subclone[colnames(srt)[srt$subclone %in% ind]])
  color_fun <- circlize::colorRamp2(breaks = c(0,2,4), colors = c("blue", "white", "red"))
  
  ht <- Heatmap(t(cnmat), cluster_columns = F, cluster_rows = F, col = color_fun, show_row_names = F, show_column_names = F, row_split = rols, 
           border =  T, name = "CN est", 
          right_annotation =  rowAnnotation(subclone = rols, col = list(subclone = setNames(ggsci::pal_futurama()(length(ind)), unique(rols))),
                                            show_annotation_name=F, annotation_legend_param = list(
                                              labels_gp = gpar(fontsize = 12),
                                            title_gp = gpar(fontsize = 16)
                                            )),
          row_title_gp = gpar(fontsize = 16),
          column_title_gp = gpar(fontsize= 12),
          heatmap_legend_param = list(labels_gp = gpar(fontsize = 12), title_gp = gpar(fontsize = 16)))
  ht
pdf("CA35_bladder_subclone_hm.pdf", width = 4, height = 4)
ht
dev.off()
```


##PSMA variants 

```{r}
data <- readRDS("~/CASCADEpaper/paper/Fig3_genomics/point_mutation/point_mutation_table_intersection.Rds")
data <-rbindlist(data, idcol = "sample")
range(as.numeric(data$tumor_AF[data$Existing_variation == "rs75111588" & grepl("CA0046", data$sample)]))
range(as.numeric(data$tumor_AF[data$Existing_variation == "rs75111588" & grepl("CA0076", data$sample)]))

data$varaint <- paste(data$CHROM, data$POS, data$REF, data$ALT)
data %>% filter(SYMBOL == "FOLH1") %>% filter(as.numeric(MAX_AF )<0.1 | is.na(as.numeric(MAX_AF)))-> psma
hist(as.numeric(psma$MAX_AF, breaks = 50))
psma <- psma[, c(1:25, 33,56,73:75)]
psma$variant <- paste(psma$CHROM, psma$POS, psma$REF, psma$ALT)

psma <- split(psma, psma$sample)
psma <- lapply(psma, function(x) return(split(x, x$variant)))

psma <- lapply(psma, lapply, function(cons){
  if(length(unique(cons$IMPACT)) == 1){
  temp <- unlist(strsplit(cons$Consequence, split = "&"))
  temp <- names(which.max(table(temp)))
  return(data.frame(IMPACT = unique(cons$IMPACT[cons$Consequence==temp]), consequence = temp, variant = unique(cons$variant), 
                    MAX_AF = max(as.numeric(cons$MAX_AF), na.rm = T), AF = mean(as.numeric(cons$AF), na.rm = T), Existing_variation = unique(cons$Existing_variation)))}
 else{
    ind <-  which(c("HIGH", "MODERATE", "MODIFIER") %in% cons$IMPACT)
    temp <- unlist(strsplit(cons$Consequence[cons$IMPACT == c("HIGH", "MODERATE", "MODIFIER")[min(ind)]], split = "&"))
    temp <- names(which.max(table(temp)))
    return(data.frame(IMPACT = c("HIGH", "MODERATE", "MODIFIER")[min(ind)], consequence = temp, variant = unique(cons$variant), 
                    MAX_AF = max(as.numeric(cons$MAX_AF), na.rm = T), AF = mean(as.numeric(cons$AF), na.rm = T), Existing_variation = unique(cons$Existing_variation)))
  }
  
})

psma <- rbindlist(lapply(psma, rbindlist), idcol = "sample")
psma <- distinct(psma)


saveRDS(psma, "FOLH1_SNP.Rds")
table(psma$Existing_variation)
```

```{r}
head(phenotype_meta)
xlab <- setNames(paste(phenotype_meta$patient, phenotype_meta$site,sapply(strsplit(phenotype_meta$sample, split = "_"), function(x) tail(x, 1))), phenotype_meta$sample)
phenotype_meta$xlab <-  xlab


mat <- as.data.frame.matrix(table(psma$sample, psma$consequence))
rownames(mat)[4:5] <-  c("CA0035_bladder_2" , "CA0035_paraaortic_lymph_node_1")
head(phenotype_meta)

mat <- mat[intersect(rownames(mat), unique(phenotype_meta$sample)),]
mat2 <- matrix(0, nrow = 34-nrow(mat), ncol = 4, dimnames = list(setdiff(unique(phenotype_meta$sample), rownames(mat)), colnames(mat)))
mat <- rbind(mat, mat2)
mat <- t(mat)
colnames(mat) <- xlab[colnames(mat)]

mat <- matrix(as.character(mat), nrow = nrow(mat), ncol = ncol(mat), dimnames = list(rownames(mat), colnames(mat)))

col.CN <- c("lightgrey", brewer.ylorrd(8))
names(col.CN) <- c(as.character(0:8))
CN.list <- lapply(col.CN, function(x) return(alter_graphic("rect", width = 0.9, height = 0.9, fill = x)))
lgd1 = Legend( title = "occurency", legend_gp = gpar(fill = col.CN[c(1:5, 9)]), at = c(0:4, 8))

mean <- phenotype_meta %>% group_by(xlab) %>% summarise(mean = mean(FOLH1_exp))
mean <- mean %>% arrange(mean)

col_fun = colorRamp2(c(min(mean$mean), max(mean$mean)), c("slategray2", "navy"))
column_ha <- HeatmapAnnotation(FOLH1_exp  = mean$mean, col = list(FOLH1_exp  = col_fun))
rownames(mat) <- gsub("_", " ", rownames(mat))
rownames(mat) <- stringr::str_wrap(rownames(mat), 14)

onc2 <-oncoPrint(mat[, mean$xlab], col = col.CN, alter_fun = c(background = function(...) NULL,CN.list),
        top_annotation  = column_ha, 
        show_column_names = T, 
        show_pct = F, row_names_side = "left",
        right_annotation = NULL,
        row_names_gp = gpar(fontsize = 10)
        )
draw(onc2, show_heatmap_legend = F,  ht_gap = unit(1, "mm"), annotation_legend_list = lgd1, padding = unit(c(6, 7, 2, 5), "mm"))

pdf("FOLH1_variants.pdf", width  = 6, height = 4)
draw(onc2, show_heatmap_legend = F,  ht_gap = unit(1, "mm"), annotation_legend_list = lgd1, padding = unit(c(6, 3, 2, 5), "mm"))
dev.off()
```